// touchguild/dist/lib/util/Util.js
"use strict";
/** @module Util */
Object.defineProperty(exports, "__esModule", { value: true });
exports.is = exports.Util = void 0;
const Member_1 = require("../structures/Member");
const Channel_1 = require("../structures/Channel");
const ForumThread_1 = require("../structures/ForumThread");
const Guild_1 = require("../structures/Guild");
const User_1 = require("../structures/User");
const GuildRole_1 = require("../structures/GuildRole");
const GuildGroup_1 = require("../structures/GuildGroup");
const GuildSubscription_1 = require("../structures/GuildSubscription");
const GuildCategory_1 = require("../structures/GuildCategory");
class Util {
    #client;
    constructor(client) {
        this.#client = client;
    }
    updateUser(user) {
        return this.#client.users.has(user.id) ? this.#client.users.update(user) : this.#client.users.add(new User_1.User(user, this.#client));
    }
    updateMember(guildID, memberID, member) {
        const guild = this.#client.guilds.get(guildID);
        if (guild && this.#client.user?.id === memberID) {
            if (guild["_clientMember"]) {
                guild["_clientMember"]["update"](member);
            }
            else {
                guild["_clientMember"] = guild.members.update({ ...member, id: memberID }, guildID);
            }
            return guild["_clientMember"];
        }
        return guild ? guild.members.update({ ...member, id: memberID }, guildID) : new Member_1.Member({ ...member }, this.#client, guildID);
    }
    updateForumThread(data) {
        if (data.serverId) {
            const guild = this.#client.guilds.get(data.serverId);
            const channel = guild?.channels.get(data.channelId);
            if (guild && channel) {
                const thread = channel.threads.has(data.id) ? channel.threads.update(data) : channel.threads.add(new ForumThread_1.ForumThread(data, this.#client));
                return thread;
            }
        }
        return new ForumThread_1.ForumThread(data, this.#client);
    }
    updateGuild(data) {
        if (data.id) {
            return this.#client.guilds.has(data.id) ? this.#client.guilds.update(data) : this.#client.guilds.add(new Guild_1.Guild(data, this.#client));
        }
        return new Guild_1.Guild(data, this.#client);
    }
    updateRole(data) {
        if (data.serverId) {
            const guild = this.#client.guilds.get(data.serverId);
            if (guild) {
                const role = guild.roles.has(data.id) ? guild.roles.update(data) : guild.roles.add(new GuildRole_1.GuildRole(data, this.#client));
                return role;
            }
        }
        return new GuildRole_1.GuildRole(data, this.#client);
    }
    updateGuildGroup(data) {
        if (data.serverId) {
            const guild = this.#client.guilds.get(data.serverId);
            if (guild) {
                const group = guild.groups.has(data.id) ? guild.groups.update(data) : guild.groups.add(new GuildGroup_1.GuildGroup(data, this.#client));
                return group;
            }
        }
        return new GuildGroup_1.GuildGroup(data, this.#client);
    }
    updateChannel(data) {
        if (data.serverId) {
            const guild = this.#client.guilds.get(data.serverId);
            if (guild) {
                const channel = guild.channels.has(data.id) ? guild.channels.update(data) : guild.channels.add(Channel_1.Channel.from(data, this.#client));
                return channel;
            }
        }
        return Channel_1.Channel.from(data, this.#client);
    }
    updateGuildSubscription(data) {
        return new GuildSubscription_1.GuildSubscription(data, this.#client);
    }
    updateGuildCategory(data) {
        return new GuildCategory_1.GuildCategory(data, this.#client);
    }
}
exports.Util = Util;
function is(input) {
    return true;
}
exports.is = is;
