// touchguild/dist/lib/rest/endpoints.js
"use strict";
/** REST/Endpoints */
/* eslint-disable @typescript-eslint/quotes */
/* eslint-disable @typescript-eslint/explicit-function-return-type */
/* eslint no-use-before-define: 0 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GUILD_MEMBER_ROLES = exports.GUILD_MEMBER_ROLE_XP = exports.GUILD_MEMBER_ROLE = exports.GUILD_ROLE_UPDATE_PERMISSION = exports.GUILD_ROLE = exports.GUILD_ROLES = exports.GUILD_BANS = exports.GUILD_BAN = exports.GUILD_SUBSCRIPTION = exports.GUILD_SUBSCRIPTIONS = exports.GUILD_GROUP_ALTERNATIVE = exports.GUILD_GROUP_MEMBERS = exports.GUILD_GROUP_MEMBER = exports.GUILD_GROUP = exports.GUILD_GROUPS = exports.GUILD_CATEGORY_ROLE_PERMISSIONS = exports.GUILD_CATEGORY_ROLE_PERMISSION = exports.GUILD_CATEGORY_USER_PERMISSIONS = exports.GUILD_CATEGORY_USER_PERMISSION = exports.GUILD_CHANNEL_USER_PERMISSIONS = exports.GUILD_CHANNEL_USER_PERMISSION = exports.GUILD_CHANNEL_ROLE_PERMISSIONS = exports.GUILD_CHANNEL_ROLE_PERMISSION = exports.GUILD_MEMBER_SOCIALS = exports.GUILD_MEMBERS = exports.GUILD_MEMBER_BULK_XP = exports.GUILD_MEMBER_XP = exports.GUILD_MEMBER_PERMISSION = exports.GUILD_MEMBER = exports.MEMBER_NICKNAME = exports.CHANNEL_ANNOUNCEMENT_COMMENT_EMOTE = exports.CHANNEL_ANNOUNCEMENT_COMMENT_EMOTES = exports.CHANNEL_ANNOUNCEMENT_COMMENT = exports.CHANNEL_ANNOUNCEMENT_COMMENTS = exports.CHANNEL_ANNOUNCEMENT_EMOTE = exports.CHANNEL_ANNOUNCEMENT_EMOTES = exports.CHANNEL_ANNOUNCEMENT = exports.CHANNEL_ANNOUNCEMENTS = exports.CHANNEL_MESSAGE_EMOTE = exports.CHANNEL_MESSAGE_EMOTES = exports.CHANNEL_MESSAGE_PIN = exports.CHANNEL_MESSAGE = exports.CHANNEL_MESSAGES = exports.USER_STATUS = exports.USER_SERVERS = exports.USER = exports.GUILD = exports.CHANNEL_ARCHIVE = exports.CHANNEL = exports.CHANNELS = void 0;
exports.GUILD_CATEGORY = exports.GUILD_CATEGORY_CREATE = exports.GUILD_WEBHOOK = exports.GUILD_WEBHOOKS = exports.CHANNEL_EVENT_EVENT_SERIES_ENTITY = exports.CHANNEL_EVENT_EVENT_SERIES = exports.CHANNEL_EVENT_COMMENT_EMOTE = exports.CHANNEL_EVENT_COMMENT_EMOTES = exports.CHANNEL_EVENT_COMMENT = exports.CHANNEL_EVENT_COMMENTS = exports.CHANNEL_EVENT_RSVPS = exports.CHANNEL_EVENT_RSVP = exports.CHANNEL_EVENT_EMOTE = exports.CHANNEL_EVENT_EMOTES = exports.CHANNEL_EVENT = exports.CHANNEL_EVENTS = exports.CHANNEL_DOC_COMMENT_EMOTE = exports.CHANNEL_DOC_COMMENT_EMOTES = exports.CHANNEL_DOC_COMMENT = exports.CHANNEL_DOC_COMMENTS = exports.CHANNEL_DOC_EMOTE = exports.CHANNEL_DOC_EMOTES = exports.CHANNEL_DOC = exports.CHANNEL_DOCS = exports.LIST_ITEM_COMPLETE = exports.LIST_ITEM = exports.LIST_ITEMS = exports.FORUM_TOPIC_COMMENT_EMOTE = exports.FORUM_TOPIC_COMMENT_EMOTES = exports.FORUM_TOPIC_COMMENT = exports.FORUM_TOPIC_COMMENTS = exports.FORUM_TOPIC_EMOTE = exports.FORUM_TOPIC_LOCK = exports.FORUM_TOPIC_PIN = exports.FORUM_TOPIC = exports.FORUM_TOPICS = void 0;
const CHANNELS = () => `/channels`;
exports.CHANNELS = CHANNELS;
const CHANNEL = (channelID) => `/channels/${channelID}`;
exports.CHANNEL = CHANNEL;
const CHANNEL_ARCHIVE = (channelID) => `/channels/${channelID}/archive`;
exports.CHANNEL_ARCHIVE = CHANNEL_ARCHIVE;
const GUILD = (guildID) => `/servers/${guildID}`;
exports.GUILD = GUILD;
const USER = (userID) => `/users/${userID}`;
exports.USER = USER;
const USER_SERVERS = (userID) => `/users/${userID}/servers`;
exports.USER_SERVERS = USER_SERVERS;
const USER_STATUS = (userID) => `/users/${userID}/status`;
exports.USER_STATUS = USER_STATUS;
const CHANNEL_MESSAGES = (channelID) => `/channels/${channelID}/messages`;
exports.CHANNEL_MESSAGES = CHANNEL_MESSAGES;
const CHANNEL_MESSAGE = (channelID, messageID) => `/channels/${channelID}/messages/${messageID}`;
exports.CHANNEL_MESSAGE = CHANNEL_MESSAGE;
const CHANNEL_MESSAGE_PIN = (channelID, messageID) => `/channels/${channelID}/messages/${messageID}/pin`;
exports.CHANNEL_MESSAGE_PIN = CHANNEL_MESSAGE_PIN;
const CHANNEL_MESSAGE_EMOTES = (channelID, messageID) => `/channels/${channelID}/messages/${messageID}/emotes`;
exports.CHANNEL_MESSAGE_EMOTES = CHANNEL_MESSAGE_EMOTES;
const CHANNEL_MESSAGE_EMOTE = (channelID, messageID, emoteID) => `/channels/${channelID}/messages/${messageID}/emotes/${emoteID}`;
exports.CHANNEL_MESSAGE_EMOTE = CHANNEL_MESSAGE_EMOTE;
const CHANNEL_ANNOUNCEMENTS = (channelID) => `/channels/${channelID}/announcements`;
exports.CHANNEL_ANNOUNCEMENTS = CHANNEL_ANNOUNCEMENTS;
const CHANNEL_ANNOUNCEMENT = (channelID, announcementID) => `/channels/${channelID}/announcements/${announcementID}`;
exports.CHANNEL_ANNOUNCEMENT = CHANNEL_ANNOUNCEMENT;
const CHANNEL_ANNOUNCEMENT_EMOTES = (channelID, announcementID) => `/channels/${channelID}/announcements/${announcementID}/emotes`;
exports.CHANNEL_ANNOUNCEMENT_EMOTES = CHANNEL_ANNOUNCEMENT_EMOTES;
const CHANNEL_ANNOUNCEMENT_EMOTE = (channelID, announcementID, emoteID) => `/channels/${channelID}/announcements/${announcementID}/emotes/${emoteID}`;
exports.CHANNEL_ANNOUNCEMENT_EMOTE = CHANNEL_ANNOUNCEMENT_EMOTE;
const CHANNEL_ANNOUNCEMENT_COMMENTS = (channelID, announcementID) => `/channels/${channelID}/announcements/${announcementID}/comments`;
exports.CHANNEL_ANNOUNCEMENT_COMMENTS = CHANNEL_ANNOUNCEMENT_COMMENTS;
const CHANNEL_ANNOUNCEMENT_COMMENT = (channelID, announcementID, commentID) => `/channels/${channelID}/announcements/${announcementID}/comments/${commentID}`;
exports.CHANNEL_ANNOUNCEMENT_COMMENT = CHANNEL_ANNOUNCEMENT_COMMENT;
const CHANNEL_ANNOUNCEMENT_COMMENT_EMOTES = (channelID, announcementID, commentID) => `/channels/${channelID}/announcements/${announcementID}/comments/${commentID}/emotes`;
exports.CHANNEL_ANNOUNCEMENT_COMMENT_EMOTES = CHANNEL_ANNOUNCEMENT_COMMENT_EMOTES;
const CHANNEL_ANNOUNCEMENT_COMMENT_EMOTE = (channelID, announcementID, commentID, emoteID) => `/channels/${channelID}/announcements/${announcementID}/comments/${commentID}/emotes/${emoteID}`;
exports.CHANNEL_ANNOUNCEMENT_COMMENT_EMOTE = CHANNEL_ANNOUNCEMENT_COMMENT_EMOTE;
const MEMBER_NICKNAME = (guildID, memberID) => `/servers/${guildID}/members/${memberID}/nickname`;
exports.MEMBER_NICKNAME = MEMBER_NICKNAME;
const GUILD_MEMBER = (guildID, memberID) => `/servers/${guildID}/members/${memberID}`;
exports.GUILD_MEMBER = GUILD_MEMBER;
const GUILD_MEMBER_PERMISSION = (guildID, memberID) => `/servers/${guildID}/members/${memberID}/permissions`;
exports.GUILD_MEMBER_PERMISSION = GUILD_MEMBER_PERMISSION;
const GUILD_MEMBER_XP = (guildID, memberID) => `/servers/${guildID}/members/${memberID}/xp`;
exports.GUILD_MEMBER_XP = GUILD_MEMBER_XP;
const GUILD_MEMBER_BULK_XP = (guildID) => `/servers/${guildID}/xp`;
exports.GUILD_MEMBER_BULK_XP = GUILD_MEMBER_BULK_XP;
const GUILD_MEMBERS = (guildID) => `/servers/${guildID}/members`;
exports.GUILD_MEMBERS = GUILD_MEMBERS;
const GUILD_MEMBER_SOCIALS = (guildID, memberID, type) => `/servers/${guildID}/members/${memberID}/social-links/${type}`;
exports.GUILD_MEMBER_SOCIALS = GUILD_MEMBER_SOCIALS;
const GUILD_CHANNEL_ROLE_PERMISSION = (serverID, channelID, roleID) => `/servers/${serverID}/channels/${channelID}/permissions/roles/${roleID}`;
exports.GUILD_CHANNEL_ROLE_PERMISSION = GUILD_CHANNEL_ROLE_PERMISSION;
const GUILD_CHANNEL_ROLE_PERMISSIONS = (serverID, channelID) => `/servers/${serverID}/channels/${channelID}/permissions/roles`;
exports.GUILD_CHANNEL_ROLE_PERMISSIONS = GUILD_CHANNEL_ROLE_PERMISSIONS;
const GUILD_CHANNEL_USER_PERMISSION = (serverID, channelID, userID) => `/servers/${serverID}/channels/${channelID}/permissions/users/${userID}`;
exports.GUILD_CHANNEL_USER_PERMISSION = GUILD_CHANNEL_USER_PERMISSION;
const GUILD_CHANNEL_USER_PERMISSIONS = (serverID, channelID) => `/servers/${serverID}/channels/${channelID}/permissions/users`;
exports.GUILD_CHANNEL_USER_PERMISSIONS = GUILD_CHANNEL_USER_PERMISSIONS;
const GUILD_CATEGORY_USER_PERMISSION = (serverID, categoryID, userID) => `/servers/${serverID}/categories/${categoryID}/permissions/users/${userID}`;
exports.GUILD_CATEGORY_USER_PERMISSION = GUILD_CATEGORY_USER_PERMISSION;
const GUILD_CATEGORY_USER_PERMISSIONS = (serverID, categoryID) => `/servers/${serverID}/categories/${categoryID}/permissions/users`;
exports.GUILD_CATEGORY_USER_PERMISSIONS = GUILD_CATEGORY_USER_PERMISSIONS;
const GUILD_CATEGORY_ROLE_PERMISSION = (serverID, categoryID, roleID) => `/servers/${serverID}/categories/${categoryID}/permissions/roles/${roleID}`;
exports.GUILD_CATEGORY_ROLE_PERMISSION = GUILD_CATEGORY_ROLE_PERMISSION;
const GUILD_CATEGORY_ROLE_PERMISSIONS = (serverID, categoryID) => `/servers/${serverID}/categories/${categoryID}/permissions/roles`;
exports.GUILD_CATEGORY_ROLE_PERMISSIONS = GUILD_CATEGORY_ROLE_PERMISSIONS;
const GUILD_GROUPS = (guildID) => `/servers/${guildID}/groups`;
exports.GUILD_GROUPS = GUILD_GROUPS;
const GUILD_GROUP = (guildID, groupID) => `/servers/${guildID}/groups/${groupID}`;
exports.GUILD_GROUP = GUILD_GROUP;
const GUILD_GROUP_MEMBER = (groupID, memberID) => `/groups/${groupID}/members/${memberID}`;
exports.GUILD_GROUP_MEMBER = GUILD_GROUP_MEMBER;
const GUILD_GROUP_MEMBERS = (groupID) => `/groups/${groupID}/members/`;
exports.GUILD_GROUP_MEMBERS = GUILD_GROUP_MEMBERS;
const GUILD_GROUP_ALTERNATIVE = (groupID) => `/groups/${groupID}`;
exports.GUILD_GROUP_ALTERNATIVE = GUILD_GROUP_ALTERNATIVE;
const GUILD_SUBSCRIPTIONS = (guildID) => `/servers/${guildID}/subscriptions/tiers`;
exports.GUILD_SUBSCRIPTIONS = GUILD_SUBSCRIPTIONS;
const GUILD_SUBSCRIPTION = (guildID, subscriptionID) => `/servers/${guildID}/subscriptions/tiers/${subscriptionID}`;
exports.GUILD_SUBSCRIPTION = GUILD_SUBSCRIPTION;
const GUILD_BAN = (guildID, memberID) => `/servers/${guildID}/bans/${memberID}`;
exports.GUILD_BAN = GUILD_BAN;
const GUILD_BANS = (guildID) => `/servers/${guildID}/bans`;
exports.GUILD_BANS = GUILD_BANS;
const GUILD_ROLES = (guildID) => `/servers/${guildID}/roles`;
exports.GUILD_ROLES = GUILD_ROLES;
const GUILD_ROLE = (guildID, roleID) => `/servers/${guildID}/roles/${roleID}`;
exports.GUILD_ROLE = GUILD_ROLE;
const GUILD_ROLE_UPDATE_PERMISSION = (guildID, roleID) => `/servers/${guildID}/roles/${roleID}/permissions`;
exports.GUILD_ROLE_UPDATE_PERMISSION = GUILD_ROLE_UPDATE_PERMISSION;
const GUILD_MEMBER_ROLE = (guildID, memberID, roleID) => `/servers/${guildID}/members/${memberID}/roles/${roleID}`;
exports.GUILD_MEMBER_ROLE = GUILD_MEMBER_ROLE;
const GUILD_MEMBER_ROLE_XP = (guildID, roleID) => `/servers/${guildID}/roles/${roleID}/xp`;
exports.GUILD_MEMBER_ROLE_XP = GUILD_MEMBER_ROLE_XP;
const GUILD_MEMBER_ROLES = (guildID, memberID) => `/servers/${guildID}/members/${memberID}/roles`;
exports.GUILD_MEMBER_ROLES = GUILD_MEMBER_ROLES;
const FORUM_TOPICS = (channelID) => `/channels/${channelID}/topics`;
exports.FORUM_TOPICS = FORUM_TOPICS;
const FORUM_TOPIC = (channelID, topicID) => `/channels/${channelID}/topics/${topicID}`;
exports.FORUM_TOPIC = FORUM_TOPIC;
const FORUM_TOPIC_PIN = (channelID, topicID) => `/channels/${channelID}/topics/${topicID}/pin`;
exports.FORUM_TOPIC_PIN = FORUM_TOPIC_PIN;
const FORUM_TOPIC_LOCK = (channelID, topicID) => `/channels/${channelID}/topics/${topicID}/lock`;
exports.FORUM_TOPIC_LOCK = FORUM_TOPIC_LOCK;
const FORUM_TOPIC_EMOTE = (channelID, topicID, emoteID) => `/channels/${channelID}/topics/${topicID}/emotes/${emoteID}`;
exports.FORUM_TOPIC_EMOTE = FORUM_TOPIC_EMOTE;
const FORUM_TOPIC_COMMENTS = (channelID, topicID) => `/channels/${channelID}/topics/${topicID}/comments`;
exports.FORUM_TOPIC_COMMENTS = FORUM_TOPIC_COMMENTS;
const FORUM_TOPIC_COMMENT = (channelID, topicID, commentID) => `/channels/${channelID}/topics/${topicID}/comments/${commentID}`;
exports.FORUM_TOPIC_COMMENT = FORUM_TOPIC_COMMENT;
const FORUM_TOPIC_COMMENT_EMOTES = (channelID, topicID, commentID) => `/channels/${channelID}/topics/${topicID}/comments/${commentID}/emotes`;
exports.FORUM_TOPIC_COMMENT_EMOTES = FORUM_TOPIC_COMMENT_EMOTES;
const FORUM_TOPIC_COMMENT_EMOTE = (channelID, topicID, commentID, emoteID) => `/channels/${channelID}/topics/${topicID}/comments/${commentID}/emotes/${emoteID}`;
exports.FORUM_TOPIC_COMMENT_EMOTE = FORUM_TOPIC_COMMENT_EMOTE;
const LIST_ITEMS = (channelID) => `/channels/${channelID}/items`;
exports.LIST_ITEMS = LIST_ITEMS;
const LIST_ITEM = (channelID, itemID) => `/channels/${channelID}/items/${itemID}`;
exports.LIST_ITEM = LIST_ITEM;
const LIST_ITEM_COMPLETE = (channelID, itemID) => `/channels/${channelID}/items/${itemID}/complete`;
exports.LIST_ITEM_COMPLETE = LIST_ITEM_COMPLETE;
const CHANNEL_DOCS = (channelID) => `/channels/${channelID}/docs`;
exports.CHANNEL_DOCS = CHANNEL_DOCS;
const CHANNEL_DOC = (channelID, docID) => `/channels/${channelID}/docs/${docID}`;
exports.CHANNEL_DOC = CHANNEL_DOC;
const CHANNEL_DOC_EMOTES = (channelID, docID) => `/channels/${channelID}/docs/${docID}/emotes`;
exports.CHANNEL_DOC_EMOTES = CHANNEL_DOC_EMOTES;
const CHANNEL_DOC_EMOTE = (channelID, docID, emoteID) => `/channels/${channelID}/docs/${docID}/emote/${emoteID}`;
exports.CHANNEL_DOC_EMOTE = CHANNEL_DOC_EMOTE;
const CHANNEL_DOC_COMMENTS = (channelID, docID) => `/channels/${channelID}/docs/${docID}/comments`;
exports.CHANNEL_DOC_COMMENTS = CHANNEL_DOC_COMMENTS;
const CHANNEL_DOC_COMMENT = (channelID, docID, commentID) => `/channels/${channelID}/docs/${docID}/comments/${commentID}`;
exports.CHANNEL_DOC_COMMENT = CHANNEL_DOC_COMMENT;
const CHANNEL_DOC_COMMENT_EMOTES = (channelID, docID, commentID) => `/channels/${channelID}/docs/${docID}/comments/${commentID}/emotes`;
exports.CHANNEL_DOC_COMMENT_EMOTES = CHANNEL_DOC_COMMENT_EMOTES;
const CHANNEL_DOC_COMMENT_EMOTE = (channelID, docID, commentID, emoteID) => `/channels/${channelID}/docs/${docID}/comments/${commentID}/emotes/${emoteID}`;
exports.CHANNEL_DOC_COMMENT_EMOTE = CHANNEL_DOC_COMMENT_EMOTE;
const CHANNEL_EVENTS = (channelID) => `/channels/${channelID}/events`;
exports.CHANNEL_EVENTS = CHANNEL_EVENTS;
const CHANNEL_EVENT = (channelID, eventID) => `/channels/${channelID}/events/${eventID}`;
exports.CHANNEL_EVENT = CHANNEL_EVENT;
const CHANNEL_EVENT_EMOTES = (channelID, eventID) => `/channels/${channelID}/events/${eventID}/emotes`;
exports.CHANNEL_EVENT_EMOTES = CHANNEL_EVENT_EMOTES;
const CHANNEL_EVENT_EMOTE = (channelID, eventID, emoteID) => `/channels/${channelID}/events/${eventID}/emotes/${emoteID}`;
exports.CHANNEL_EVENT_EMOTE = CHANNEL_EVENT_EMOTE;
const CHANNEL_EVENT_RSVP = (channelID, eventID, memberID) => `/channels/${channelID}/events/${eventID}/rsvps/${memberID}`;
exports.CHANNEL_EVENT_RSVP = CHANNEL_EVENT_RSVP;
const CHANNEL_EVENT_RSVPS = (channelID, eventID) => `/channels/${channelID}/events/${eventID}/rsvps`;
exports.CHANNEL_EVENT_RSVPS = CHANNEL_EVENT_RSVPS;
const CHANNEL_EVENT_COMMENTS = (channelID, eventID) => `/channels/${channelID}/events/${eventID}/comments`;
exports.CHANNEL_EVENT_COMMENTS = CHANNEL_EVENT_COMMENTS;
const CHANNEL_EVENT_COMMENT = (channelID, eventID, commentID) => `/channels/${channelID}/events/${eventID}/comments/${commentID}`;
exports.CHANNEL_EVENT_COMMENT = CHANNEL_EVENT_COMMENT;
const CHANNEL_EVENT_COMMENT_EMOTES = (channelID, eventID, commentID) => `/channels/${channelID}/events/${eventID}/comments/${commentID}/emotes`;
exports.CHANNEL_EVENT_COMMENT_EMOTES = CHANNEL_EVENT_COMMENT_EMOTES;
const CHANNEL_EVENT_COMMENT_EMOTE = (channelID, eventID, commentID, emoteID) => `/channels/${channelID}/events/${eventID}/comments/${commentID}/emotes/${emoteID}`;
exports.CHANNEL_EVENT_COMMENT_EMOTE = CHANNEL_EVENT_COMMENT_EMOTE;
const CHANNEL_EVENT_EVENT_SERIES = (channelID) => `/channels/${channelID}/event_series/`;
exports.CHANNEL_EVENT_EVENT_SERIES = CHANNEL_EVENT_EVENT_SERIES;
const CHANNEL_EVENT_EVENT_SERIES_ENTITY = (channelID, seriesID) => `/channels/${channelID}/event_series/${seriesID}`;
exports.CHANNEL_EVENT_EVENT_SERIES_ENTITY = CHANNEL_EVENT_EVENT_SERIES_ENTITY;
const GUILD_WEBHOOKS = (guildID) => `/servers/${guildID}/webhooks`;
exports.GUILD_WEBHOOKS = GUILD_WEBHOOKS;
const GUILD_WEBHOOK = (guildID, webhookID) => `/servers/${guildID}/webhooks/${webhookID}`;
exports.GUILD_WEBHOOK = GUILD_WEBHOOK;
const GUILD_CATEGORY_CREATE = (guildID) => `/servers/${guildID}/categories`;
exports.GUILD_CATEGORY_CREATE = GUILD_CATEGORY_CREATE;
const GUILD_CATEGORY = (guildID, categoryID) => `/servers/${guildID}/categories/${categoryID}`;
exports.GUILD_CATEGORY = GUILD_CATEGORY;
