// touchguild/dist/lib/rest/RESTManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RESTManager = void 0;
/** @module RESTManager */
const RequestHandler_1 = require("./RequestHandler");
const Guilds_1 = require("../routes/Guilds");
const Channels_1 = require("../routes/Channels");
const Misc_1 = require("../routes/Misc");
class RESTManager {
    #ws;
    constructor(client, options) {
        this.#ws = client.ws;
        this.client = client;
        this.token = this.#ws.token;
        this.handler = new RequestHandler_1.RequestHandler(this, options);
        this.guilds = new Guilds_1.Guilds(this);
        this.channels = new Channels_1.Channels(this);
        this.misc = new Misc_1.Miscellaneous(this);
    }
    /** Send an authenticated request.
     * @param options Request options.
     */
    async authRequest(options) {
        return this.handler.authRequest(options);
    }
    /** Send a request.
     * @param options Request options.
     */
    async request(options) {
        return this.handler.request(options);
    }
}
exports.RESTManager = RESTManager;
