// touchguild/dist/lib/structures/AnnouncementComment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnnouncementComment = void 0;
const Base_1 = require("./Base");
/** AnnouncementComment represents a comment from an Announcement channel. */
class AnnouncementComment extends Base_1.Base {
    /**
     * @param data raw data.
     * @param client client.
     */
    constructor(data, client, options) {
        super(data.id, client);
        this.content = data.content;
        this.createdAt = new Date(data.createdAt);
        this.editedTimestamp = data.updatedAt ? new Date(data.updatedAt) : null;
        this.memberID = data.createdBy;
        this.channelID = data.channelId;
        this.announcementID = data.announcementId;
        this.mentions = data.mentions ?? null;
        this.guildID = options?.guildID ?? null;
        this.update(data);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            content: this.content,
            createdAt: this.createdAt,
            editedTimestamp: this.editedTimestamp,
            memberID: this.memberID,
            channelID: this.channelID,
            announcementID: this.announcementID,
            mentions: this.mentions,
            guildID: this.guildID
        };
    }
    update(data) {
        if (data.announcementId !== undefined) {
            this.announcementID = data.announcementId;
        }
        if (data.channelId !== undefined) {
            this.channelID = data.channelId;
        }
        if (data.content !== undefined) {
            this.content = data.content;
        }
        if (data.createdAt !== undefined) {
            this.createdAt = new Date(data.createdAt);
        }
        if (data.createdBy !== undefined) {
            this.memberID = data.createdBy;
        }
        if (data.id !== undefined) {
            this.id = data.id;
        }
        if (data.mentions !== undefined) {
            this.mentions = data.mentions;
        }
        if (data.updatedAt !== undefined) {
            this.editedTimestamp = new Date(data.updatedAt);
        }
    }
    /** Retrieve the member who sent this comment, if cached.
     * If there is no cached member, this will make a rest request which returns a Promise.
     * If the request fails, it'll return undefined or throw an error that you can catch.
     */
    get member() {
        if (this.guildID === null)
            throw new Error("Couldn't get member because API didn't return guildID.");
        return (this.client.getGuild(this.guildID)?.members.get(this.memberID) ?? this.guildID) ? this.client.rest.guilds.getMember(this.guildID, this.memberID) : undefined;
    }
    /**
     * Edit this comment.
     * @param options Edit options
     */
    async edit(options) {
        return this.client.rest.channels.editAnnouncementComment(this.channelID, this.announcementID, this.id, options);
    }
    /**
     * Delete this comment.
     */
    async delete() {
        return this.client.rest.channels.deleteAnnouncementComment(this.channelID, this.announcementID, this.id);
    }
    /**
     * Create a comment in the same announcement as this one.
     * @param options Create options.
     */
    async createComment(options) {
        return this.client.rest.channels.createAnnouncementComment(this.channelID, this.announcementID, options);
    }
    /**
     * Add a reaction to this comment.
     * @param reactionID ID of the emote to add
     */
    async createReaction(emoteID) {
        return this.client.rest.channels.createReactionToSubcategory(this.channelID, "AnnouncementComment", this.announcementID, this.id, emoteID);
    }
    /**
     * Remove a reaction from this comment.
     * @param reactionID ID of the emote to remove
     */
    async deleteReaction(emoteID) {
        return this.client.rest.channels.deleteReactionFromSubcategory(this.channelID, "AnnouncementComment", this.announcementID, this.id, emoteID);
    }
}
exports.AnnouncementComment = AnnouncementComment;
