// touchguild/dist/lib/structures/Announcement.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Announcement = void 0;
const tslib_1 = require("tslib");
const Base_1 = require("./Base");
const AnnouncementComment_1 = require("./AnnouncementComment");
const TypedCollection_1 = tslib_1.__importDefault(require("../util/TypedCollection"));
/** Represents a channel announcement. */
class Announcement extends Base_1.Base {
    /**
     * @param data raw data.
     * @param client client.
     */
    constructor(data, client) {
        super(data.id, client);
        this.guildID = data.serverId;
        this.channelID = data.channelId;
        this.createdAt = new Date(data.createdAt);
        this.memberID = data.createdBy;
        this.content = data.content;
        this.mentions = data.mentions ?? null;
        this.title = data.title;
        this.comments = new TypedCollection_1.default(AnnouncementComment_1.AnnouncementComment, client, client.params.collectionLimits?.announcementComments);
        this.update(data);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            guildID: this.guildID,
            channelID: this.channelID,
            createdAt: this.createdAt,
            memberID: this.memberID,
            content: this.content,
            mentions: this.mentions,
            title: this.title
        };
    }
    update(data) {
        if (data.channelId !== undefined) {
            this.channelID = data.channelId;
        }
        if (data.content !== undefined) {
            this.content = data.content;
        }
        if (data.createdAt !== undefined) {
            this.createdAt = new Date(data.createdAt);
        }
        if (data.createdBy !== undefined) {
            this.memberID = data.createdBy;
        }
        if (data.id !== undefined) {
            this.id = data.id;
        }
        if (data.mentions !== undefined) {
            this.mentions = data.mentions;
        }
        if (data.serverId !== undefined) {
            this.guildID = data.serverId;
        }
        if (data.title !== undefined) {
            this.title = data.title;
        }
    }
    /**
     * Edit this announcement.
     * @param options Edit options
     */
    async edit(options) {
        return this.client.rest.channels.editAnnouncement(this.channelID, this.id, options);
    }
    /**
     * Delete this announcement.
     */
    async delete() {
        return this.client.rest.channels.deleteAnnouncement(this.channelID, this.id);
    }
    /**
     * Create an announcement in the same Announcement channel as this one.
     * @param options Create options.
     */
    async createAnnouncement(options) {
        return this.client.rest.channels.createAnnouncement(this.channelID, options);
    }
    /**
     * Add a reaction to this announcement.
     * @param reactionID ID of the emote to add
     */
    async createReaction(emoteID) {
        return this.client.rest.channels.createReaction(this.channelID, "ChannelAnnouncement", this.id, emoteID);
    }
    /**
     * Remove a reaction from this announcement.
     * @param reactionID ID of the emote to remove
     */
    async deleteReaction(emoteID) {
        return this.client.rest.channels.deleteReaction(this.channelID, "ChannelAnnouncement", this.id, emoteID);
    }
}
exports.Announcement = Announcement;
