// touchguild/dist/lib/structures/AnnouncementReactionInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnnouncementReactionInfo = void 0;
/** @module AnnouncementReactionInfo */
const ReactionInfo_1 = require("./ReactionInfo");
/** Information about a Announcement's reaction. */
class AnnouncementReactionInfo extends ReactionInfo_1.ReactionInfo {
    /**
     * @param data raw data.
     * @param client client.
     */
    constructor(data, client) {
        super(data, client);
        this.announcementID = data.reaction.announcementId;
        this.commentID = data.reaction["announcementCommentId"] ?? null;
        this.type = data.reaction["announcementCommentId"] ? "comment" : "announcement";
    }
    /** The announcement where the reaction has been added.
     * If the event is cached, it'll return a Announcement component,
     * otherwise it'll return basic information about this announcement.
     */
    get announcement() {
        return this.client.getChannel(this.raw.serverId, this.raw.reaction.channelId)?.announcements.get(this.announcementID) ?? {
            id: this.announcementID,
            guild: this.client.guilds.get(this.raw.serverId) ?? {
                id: this.raw.serverId
            },
            channelID: this.raw.reaction.channelId
        };
    }
}
exports.AnnouncementReactionInfo = AnnouncementReactionInfo;
