// touchguild/dist/lib/structures/BannedMember.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BannedMember = void 0;
const User_1 = require("./User");
const Base_1 = require("./Base");
/** BannedMember represents a banned guild member. */
class BannedMember extends Base_1.Base {
    /**
     * @param guildID ID of the guild.
     * @param data raw data.
     * @param client client.
     */
    constructor(guildID, data, client) {
        super(data.user.id, client);
        this.guildID = guildID;
        this.ban = {
            reason: data.reason,
            createdAt: data.createdAt ? new Date(data.createdAt) : null,
            bannedBy: data.createdBy
        };
        this.user = client.users.update(data.user) ?? new User_1.User(data.user, client);
        this.member = client.getGuild(guildID)?.members.get(data.user.id) ?? null;
        this.update(data);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            guildID: this.guildID,
            ban: this.ban
        };
    }
    update(data) {
        if (data.createdAt !== undefined) {
            this.ban.createdAt = new Date(data.createdAt);
        }
        if (data.createdBy !== undefined) {
            this.ban.bannedBy = data.createdBy;
        }
        if (data.reason !== undefined) {
            this.ban.reason = data.reason;
        }
        if (data.user !== undefined && this.client.users.update(data.user)) {
            this.user = this.client.users.update(data.user);
        }
    }
    /** Getter used to get the message's guild
     *
     * Note: this can return a promise, make sure to await it before.
     */
    get guild() {
        return this.client.guilds.get(this.guildID) ?? this.client.rest.guilds.getGuild(this.guildID);
    }
}
exports.BannedMember = BannedMember;
