// touchguild/dist/lib/structures/CalendarEvent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CalendarEvent = void 0;
const tslib_1 = require("tslib");
const Base_1 = require("./Base");
const CalendarRSVP_1 = require("./CalendarRSVP");
const CalendarEventComment_1 = require("./CalendarEventComment");
const TypedCollection_1 = tslib_1.__importDefault(require("../util/TypedCollection"));
/** CalendarEvent represents an event coming from a calendar channel. */
class CalendarEvent extends Base_1.Base {
    /**
     * @param data raw data.
     * @param client client.
     */
    constructor(data, client) {
        super(data.id, client);
        this.data = data;
        this.id = data.id;
        this.guildID = data.serverId;
        this.channelID = data.channelId;
        this.name = data.name ?? null;
        this.description = data.description ?? null;
        this.location = data.location ?? null;
        this.url = data.url ?? null;
        this.color = data.color ?? null;
        this.rsvpLimit = data.rsvpLimit ?? null;
        this.startsAt = data.startsAt ? new Date(data.startsAt) : null;
        this.duration = data.duration * 60000 ?? null; // in ms.
        this.isPrivate = data.isPrivate ?? false;
        this.mentions = data.mentions ?? null;
        this.createdAt = data.createdAt ? new Date(data.createdAt) : null;
        this.ownerID = data.createdBy;
        this.cancelation = data.cancellation ?? null;
        this.rsvps = new TypedCollection_1.default(CalendarRSVP_1.CalendarEventRSVP, client, client.params.collectionLimits?.scheduledEventsRSVPS);
        this.comments = new TypedCollection_1.default(CalendarEventComment_1.CalendarEventComment, client, client.params.collectionLimits?.calendarComments);
        this.update(data);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            data: this.data,
            id: this.id,
            guildID: this.guildID,
            channelID: this.channelID,
            name: this.name,
            description: this.description,
            location: this.location,
            url: this.url,
            color: this.color,
            rsvpLimit: this.rsvpLimit,
            startsAt: this.startsAt,
            duration: this.duration,
            isPrivate: this.isPrivate,
            mentions: this.mentions,
            createdAt: this.createdAt,
            ownerID: this.ownerID,
            cancelation: this.cancelation,
            rsvps: this.rsvps.map(rsvp => rsvp.toJSON())
        };
    }
    update(data) {
        if (data.cancellation !== undefined) {
            this.cancelation = data.cancellation;
        }
        if (data.channelId !== undefined) {
            this.channelID = data.channelId;
        }
        if (data.color !== undefined) {
            this.color = data.color;
        }
        if (data.createdAt !== undefined) {
            this.createdAt = new Date(data.createdAt);
        }
        if (data.createdBy !== undefined) {
            this.ownerID = data.createdBy;
        }
        if (data.description !== undefined) {
            this.description = data.description;
        }
        if (data.duration !== undefined) {
            this.duration = data.duration;
        }
        if (data.id !== undefined) {
            this.id = data.id;
        }
        if (data.isPrivate !== undefined) {
            this.isPrivate = data.isPrivate;
        }
        if (data.location !== undefined) {
            this.location = data.location;
        }
        if (data.mentions !== undefined) {
            this.mentions = data.mentions;
        }
        if (data.name !== undefined) {
            this.name = data.name;
        }
        if (data.rsvpLimit !== undefined) {
            this.rsvpLimit = data.rsvpLimit;
        }
        if (data.serverId !== undefined) {
            this.guildID = data.serverId;
        }
        if (data.startsAt !== undefined) {
            this.startsAt = new Date(data.startsAt);
        }
        if (data.url !== undefined) {
            this.url = data.url;
        }
    }
    /** Retrieve the event's owner, if cached.
     *
     * Note: this getter can output: Member, User, Promise<Member> or undefined.
     */
    get owner() {
        const guild = this.client.guilds.get(this.guildID);
        if (guild?.members?.get(this.ownerID) && this.ownerID) {
            return guild?.members.get(this.ownerID);
        }
        else if (this.client.users.get(this.ownerID) && this.ownerID) {
            return this.client.users.get(this.ownerID);
        }
        else if (this.ownerID && this.guildID) {
            return this.client.rest.guilds.getMember(this.guildID, this.ownerID);
        }
    }
    /** Edit this event */
    async edit(options) {
        return this.client.rest.channels.editCalendarEvent(this.channelID, this.id, options);
    }
    /** Delete this event */
    async delete() {
        return this.client.rest.channels.deleteCalendarEvent(this.channelID, this.id);
    }
}
exports.CalendarEvent = CalendarEvent;
