// touchguild/dist/lib/structures/CalendarReactionInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CalendarReactionInfo = void 0;
/** @module CalendarReactionInfo */
const ReactionInfo_1 = require("./ReactionInfo");
/** Information about a CalendarEvent's reaction. */
class CalendarReactionInfo extends ReactionInfo_1.ReactionInfo {
    /**
     * @param data raw data.
     * @param client client.
     */
    constructor(data, client) {
        super(data, client);
        this.eventID = data.reaction.calendarEventId;
        this.commentID = data.reaction["calendarEventCommentId"] ?? null;
        this.type = data.reaction["calendarEventCommentId"] ? "comment" : "event";
    }
    /** The calendar event where the reaction has been added.
     * If the event is cached, it'll return a CalendarEvent component,
     * otherwise it'll return basic information about this event.
     */
    get event() {
        return this.client.getChannel(this.raw.serverId, this.raw.reaction.channelId)?.scheduledEvents.get(this.eventID) ?? {
            id: this.eventID,
            guild: this.client.guilds.get(this.raw.serverId) ?? {
                id: this.raw.serverId
            },
            channelID: this.raw.reaction.channelId
        };
    }
}
exports.CalendarReactionInfo = CalendarReactionInfo;
