// touchguild/dist/lib/structures/CalendarRSVP.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CalendarEventRSVP = void 0;
const Base_1 = require("./Base");
/** CalendarEventRSVP represents a guild member's event RSVP.
 * It gives information about a member's set presence to an event.
 */
class CalendarEventRSVP extends Base_1.Base {
    /** Raw data */
    #data;
    /**
     * @param data raw data.
     * @param client client.
     */
    constructor(data, client) {
        super(data.calendarEventId, client);
        this.#data = data;
        this.guildID = data.serverId;
        this.channelID = data.channelId;
        this.entityID = data.userId;
        this.status = data.status;
        this.creatorID = data.createdBy ?? null;
        this.updatedBy = data.updatedBy ?? null;
        this.updatedAt = data.updatedAt ? new Date(data.updatedAt) : null;
        this.createdAt = data.createdAt ? new Date(data.createdAt) : null;
        this.update(data);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            guildID: this.guildID,
            channelID: this.channelID,
            entityID: this.entityID,
            status: this.status,
            creatorID: this.creatorID,
            updatedBy: this.updatedBy,
            updatedAt: this.updatedAt,
            createdAt: this.createdAt
        };
    }
    update(data) {
        if (data.calendarEventId !== undefined) {
            this.id = data.calendarEventId;
        }
        if (data.channelId !== undefined) {
            this.channelID = data.channelId;
        }
        if (data.createdAt !== undefined) {
            this.createdAt = new Date(data.createdAt);
        }
        if (data.createdBy !== undefined) {
            this.creatorID = data.createdBy;
        }
        if (data.serverId !== undefined) {
            this.guildID = data.serverId;
        }
        if (data.status !== undefined) {
            this.status = data.status;
        }
        if (data.updatedAt !== undefined) {
            this.updatedAt = new Date(data.updatedAt);
        }
        if (data.updatedBy !== undefined) {
            this.updatedBy = data.updatedBy;
        }
        if (data.userId !== undefined) {
            this.entityID = data.userId;
        }
    }
    /** Edit this RSVP. */
    async edit(options) {
        return this.client.rest.channels.editCalendarRsvp(this.channelID, this.id, this.entityID, options);
    }
    /** Delete this RSVP. */
    async delete() {
        return this.client.rest.channels.deleteCalendarRsvp(this.channelID, this.id, this.entityID);
    }
}
exports.CalendarEventRSVP = CalendarEventRSVP;
