// touchguild/dist/lib/structures/CalendarEventComment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CalendarEventComment = void 0;
const Base_1 = require("./Base");
/** CalendarEventComment represents an event comment coming from a calendar channel. */
class CalendarEventComment extends Base_1.Base {
    /**
     * @param data raw data.
     * @param client client.
     */
    constructor(data, client, options) {
        super(data.id, client);
        this.data = data;
        this.guildID = options?.guildID ?? null;
        this.content = data.content;
        this.createdAt = new Date(data.createdAt);
        this.updatedAt = data.updatedAt ? new Date(data.updatedAt) : null;
        this.eventID = Number(data.calendarEventId);
        this.channelID = data.channelId;
        this.memberID = data.createdBy;
        this.update(data);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            data: this.data,
            content: this.content,
            createdAt: this.createdAt,
            updatedAt: this.updatedAt,
            eventID: this.eventID,
            channelID: this.channelID,
            memberID: this.memberID
        };
    }
    update(data) {
        if (data.calendarEventId !== undefined) {
            this.eventID = Number(data.calendarEventId);
        }
        if (data.channelId !== undefined) {
            this.channelID = data.channelId;
        }
        if (data.content !== undefined) {
            this.content = data.content;
        }
        if (data.createdAt !== undefined) {
            this.createdAt = new Date(this.createdAt);
        }
        if (data.createdBy !== undefined) {
            this.memberID = data.createdBy;
        }
        if (data.id !== undefined) {
            this.id = data.id;
        }
        if (data.updatedAt !== undefined) {
            this.updatedAt = data.updatedAt ? new Date(data.updatedAt) : null;
        }
    }
    /** Retrieve the member who sent this comment, if cached.
     * If there is no cached member, this will make a rest request which returns a Promise.
     * If the request fails, it'll return undefined or throw an error that you can catch.
     */
    get member() {
        if (this.guildID === null)
            throw new Error("Couldn't get member because API didn't return guildID.");
        return (this.client.getGuild(this.guildID)?.members.get(this.memberID) ?? this.guildID) ? this.client.rest.guilds.getMember(this.guildID, this.memberID) : undefined;
    }
    /** Create a comment in the same event as this one.
     * @param options Create options.
     */
    async createCalendarComment(options) {
        return this.client.rest.channels.createCalendarComment(this.channelID, this.eventID, options);
    }
    /** Add a reaction to this comment.
     * @param reaction ID of the reaction to add.
     */
    async createReaction(reaction) {
        return this.client.rest.channels.createReactionToSubcategory(this.channelID, "CalendarEventComment", this.eventID, this.id, reaction);
    }
    /** Remove a reaction from this comment.
     * @param reaction ID of the reaction to remove.
     */
    async deleteReaction(reaction) {
        return this.client.rest.channels.deleteReactionFromSubcategory(this.channelID, "CalendarEventComment", this.eventID, this.id, reaction);
    }
    /** Edit this comment */
    async edit(options) {
        return this.client.rest.channels.editCalendarComment(this.channelID, this.eventID, this.id, options);
    }
    /** Delete this comment */
    async delete() {
        return this.client.rest.channels.deleteCalendarComment(this.channelID, this.eventID, this.id);
    }
}
exports.CalendarEventComment = CalendarEventComment;
