// touchguild/dist/lib/structures/CalendarChannel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CalendarChannel = void 0;
const tslib_1 = require("tslib");
const CalendarEvent_1 = require("./CalendarEvent");
const GuildChannel_1 = require("./GuildChannel");
const TypedCollection_1 = tslib_1.__importDefault(require("../util/TypedCollection"));
/** Represents a calendar channel. */
class CalendarChannel extends GuildChannel_1.GuildChannel {
    /**
     * @param data raw data
     * @param client client
     */
    constructor(data, client) {
        super(data, client);
        this.scheduledEvents = new TypedCollection_1.default(CalendarEvent_1.CalendarEvent, client, client.params.collectionLimits?.scheduledEvents);
        this.update(data);
    }
    /** Create an event in this channel.
     * @param options Event options.
     * @param createSeries (optional) Create a series. (event's repetition)
     */
    async createEvent(options, createSeries) {
        return this.client.rest.channels.createCalendarEvent(this.id, options, createSeries);
    }
    /** Edit an event from this channel.
     * @param eventID ID of a calendar event.
     * @param options Edit options.
     */
    async editEvent(eventID, options) {
        return this.client.rest.channels.editCalendarEvent(this.id, eventID, options);
    }
    /**
     * Edit an event series set in this channel.
     * @param channelID ID of the channel.
     * @param eventID ID of the event.
     * @param seriesID ID of the series.
     * @param options Edit repetition options.
     */
    async editSeries(eventID, seriesID, options) {
        return this.client.rest.channels.editCalendarEventSeries(this.id, eventID, seriesID, options);
    }
    /**
     * Delete an event series set in this channel.
     * @param eventID ID of the event.
     * @param seriesID ID of the series.
     */
    async deleteSeries(eventID, seriesID) {
        return this.client.rest.channels.deleteCalendarEventSeries(this.id, eventID, seriesID);
    }
    /**
     * Delete an event from this channel.
     * @param eventID ID of the event to delete.
     */
    async deleteEvent(eventID) {
        return this.client.rest.channels.deleteCalendarEvent(this.id, eventID);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            scheduledEvents: this.scheduledEvents.map(event => event.toJSON())
        };
    }
}
exports.CalendarChannel = CalendarChannel;
