// touchguild/dist/lib/structures/Channel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Channel = void 0;
const Base_1 = require("./Base");
const GuildChannel_1 = require("./GuildChannel");
const CalendarChannel_1 = require("./CalendarChannel");
const DocChannel_1 = require("./DocChannel");
const ForumChannel_1 = require("./ForumChannel");
const TextChannel_1 = require("./TextChannel");
const AnnouncementChannel_1 = require("./AnnouncementChannel");
/** Represents a channel. */
class Channel extends Base_1.Base {
    /**
     * @param data raw data
     * @param client client
     */
    constructor(data, client) {
        super(data.id, client);
        this.type = data.type;
        this.name = data.name;
    }
    static from(data, client) {
        switch (data.type) {
            case "announcements": {
                return new AnnouncementChannel_1.AnnouncementChannel(data, client);
            }
            case "calendar": {
                return new CalendarChannel_1.CalendarChannel(data, client);
            }
            case "chat": {
                return new TextChannel_1.TextChannel(data, client);
            }
            case "docs": {
                return new DocChannel_1.DocChannel(data, client);
            }
            case "forums": {
                return new ForumChannel_1.ForumChannel(data, client);
            }
            case "list": {
                return new GuildChannel_1.GuildChannel(data, client);
            }
            case "media": {
                return new GuildChannel_1.GuildChannel(data, client);
            }
            case "scheduling": {
                return new GuildChannel_1.GuildChannel(data, client);
            }
            case "stream": {
                return new GuildChannel_1.GuildChannel(data, client);
            }
            case "voice": {
                return new GuildChannel_1.GuildChannel(data, client);
            }
            default: {
                return new Channel(data, client);
            }
        }
    }
    toJSON() {
        return {
            ...super.toJSON(),
            type: this.type,
            name: this.name
        };
    }
    /** Edit the channel. */
    async edit(options) {
        return this.client.rest.guilds.editChannel(this.id, options);
    }
    /** Delete the channel. */
    async delete() {
        return this.client.rest.guilds.deleteChannel(this.id);
    }
}
exports.Channel = Channel;
