// touchguild/dist/lib/structures/Doc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Doc = void 0;
const tslib_1 = require("tslib");
const Base_1 = require("./Base");
const DocComment_1 = require("./DocComment");
const TypedCollection_1 = tslib_1.__importDefault(require("../util/TypedCollection"));
/** Doc represents an item of a "Docs" channel. */
class Doc extends Base_1.Base {
    /**
     * @param data raw data
     * @param client client
     */
    constructor(data, client) {
        super(data.id, client);
        this.guildID = data.serverId;
        this.channelID = data.channelId;
        this.name = data.title ?? null;
        this.content = data.content ?? null;
        this.mentions = data.mentions ?? {};
        this.createdAt = new Date(data.createdAt);
        this.memberID = data.createdBy;
        this.editedTimestamp = data.updatedAt ? new Date(data.updatedAt) : null;
        this.updatedBy = data.updatedBy ?? null;
        this.comments = new TypedCollection_1.default(DocComment_1.DocComment, client, client.params.collectionLimits?.docComments);
        this.update(data);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            guildID: this.guildID,
            channelID: this.channelID,
            name: this.name,
            content: this.content,
            mentions: this.mentions,
            createdAt: this.createdAt,
            memberID: this.memberID,
            editedTimestamp: this.editedTimestamp,
            updatedBy: this.updatedBy
        };
    }
    update(data) {
        if (data.channelId !== undefined) {
            this.channelID = data.channelId;
        }
        if (data.content !== undefined) {
            this.content = data.content;
        }
        if (data.createdAt !== undefined) {
            this.createdAt = new Date(data.createdAt);
        }
        if (data.createdBy !== undefined) {
            this.memberID = data.createdBy;
        }
        if (data.id !== undefined) {
            this.id = data.id;
        }
        if (data.mentions !== undefined) {
            this.mentions = data.mentions;
        }
        if (data.serverId !== undefined) {
            this.guildID = data.serverId;
        }
        if (data.title !== undefined) {
            this.name = data.title;
        }
        if (data.updatedAt !== undefined) {
            this.editedTimestamp = new Date(data.updatedAt);
        }
        if (data.updatedBy !== undefined) {
            this.updatedBy = data.updatedBy;
        }
    }
    /** Retrieve the member who executed this action.
     * Note: If this doc has been edited, the updatedBy id will be used to get you the member.
     */
    get member() {
        return this.client.getGuild(this.guildID)?.members.get(this.updatedBy ?? this.memberID) ?? this.client.rest.guilds.getMember(this.guildID, this.updatedBy ?? this.memberID);
    }
    /** Edit this doc.
     * @param options Edit options.
     */
    async edit(options) {
        return this.client.rest.channels.editDoc(this.channelID, this.id, options);
    }
    /** Delete this doc. */
    async delete() {
        return this.client.rest.channels.deleteDoc(this.channelID, this.id);
    }
}
exports.Doc = Doc;
