// touchguild/dist/lib/structures/DocChannel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DocChannel = void 0;
const tslib_1 = require("tslib");
const Doc_1 = require("./Doc");
const GuildChannel_1 = require("./GuildChannel");
const TypedCollection_1 = tslib_1.__importDefault(require("../util/TypedCollection"));
/** Represents a "docs" channel. */
class DocChannel extends GuildChannel_1.GuildChannel {
    /**
     * @param data raw data
     * @param client client
     */
    constructor(data, client) {
        super(data, client);
        this.docs = new TypedCollection_1.default(Doc_1.Doc, client, client.params.collectionLimits?.docs);
        this.update(data);
    }
    /** Create a doc in this channel.
     * @param options Doc's options.
     */
    async createDoc(options) {
        return this.client.rest.channels.createDoc(this.id, options);
    }
    /** Edit a doc from this channel.
     * @param docID ID of a doc.
     * @param options Edit options.
     */
    async editDoc(docID, options) {
        return this.client.rest.channels.editDoc(this.id, docID, options);
    }
    /** Delete a doc from this channel.
     * @param docID ID of a doc.
     */
    async deleteDoc(docID) {
        return this.client.rest.channels.deleteDoc(this.id, docID);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            docs: this.docs.map(doc => doc.toJSON())
        };
    }
}
exports.DocChannel = DocChannel;
