// touchguild/dist/lib/structures/DocComment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DocComment = void 0;
const Base_1 = require("./Base");
/** DocComment represents a doc comment coming from a Docs channel. */
class DocComment extends Base_1.Base {
    /**
     * @param data raw data.
     * @param client client.
     */
    constructor(data, client, options) {
        super(data.id, client);
        this.raw = data;
        this.content = data.content;
        this.createdAt = new Date(data.createdAt);
        this.memberID = data.createdBy;
        this.updatedAt = data.updatedAt ? new Date(data.updatedAt) : null;
        this.channelID = data.channelId;
        this.docID = data.docId;
        this.mentions = data.mentions ?? null;
        this.guildID = options?.guildID ?? null;
        this.update(data);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            raw: this.raw,
            content: this.content,
            createdAt: this.createdAt,
            memberID: this.memberID,
            updatedAt: this.updatedAt,
            channelID: this.channelID,
            docID: this.docID,
            mentions: this.mentions,
            guildID: this.guildID
        };
    }
    update(data) {
        if (data.channelId !== undefined) {
            this.channelID = data.channelId;
        }
        if (data.content !== undefined) {
            this.content = data.content;
        }
        if (data.createdAt !== undefined) {
            this.createdAt = new Date(data.createdAt);
        }
        if (data.createdBy !== undefined) {
            this.memberID = data.createdBy;
        }
        if (data.docId !== undefined) {
            this.docID = data.docId;
        }
        if (data.id !== undefined) {
            this.id = data.id;
        }
        if (data.mentions !== undefined) {
            this.mentions = data.mentions;
        }
        if (data.updatedAt !== undefined) {
            this.updatedAt = new Date(data.updatedAt);
        }
    }
    /** Retrieve the member who sent this comment, if cached.
     * If there is no cached member, this will make a rest request which returns a Promise.
     * If the request fails, it'll return undefined or throw an error that you can catch.
     */
    get member() {
        if (this.guildID === null)
            throw new Error("Couldn't get member because API didn't return guildID.");
        return (this.client.getGuild(this.guildID)?.members.get(this.memberID) ?? this.guildID) ? this.client.rest.guilds.getMember(this.guildID, this.memberID) : undefined;
    }
    /** Create a comment in the same doc as this one.
     * @param options Create options.
     */
    async createDocComment(options) {
        return this.client.rest.channels.createDocComment(this.channelID, this.docID, options);
    }
    /** Add a reaction to this comment.
     * @param reaction ID of the reaction to add.
     */
    async createReaction(reaction) {
        return this.client.rest.channels.createReactionToSubcategory(this.channelID, "DocComment", this.docID, this.id, reaction);
    }
    /** Remove a reaction from this comment.
     * @param reaction ID of the reaction to remove.
     */
    async deleteReaction(reaction) {
        return this.client.rest.channels.deleteReactionFromSubcategory(this.channelID, "DocComment", this.docID, this.id, reaction);
    }
    /** Edit this comment */
    async edit(options) {
        return this.client.rest.channels.editDocComment(this.channelID, this.docID, this.id, options);
    }
    /** Delete this comment */
    async delete() {
        return this.client.rest.channels.deleteDocComment(this.channelID, this.docID, this.id);
    }
}
exports.DocComment = DocComment;
