// touchguild/dist/lib/structures/DocReactionInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DocReactionInfo = void 0;
/** @module DocReactionInfo */
const ReactionInfo_1 = require("./ReactionInfo");
/** Information about a Doc's reaction. */
class DocReactionInfo extends ReactionInfo_1.ReactionInfo {
    /**
     * @param data raw data.
     * @param client client.
     */
    constructor(data, client) {
        super(data, client);
        this.docID = data.reaction.docId;
        this.type = data["docCommentId"] ? "comment" : "doc";
    }
    /** The doc where the reaction has been added.
     * If the doc is cached, it'll return a Doc component,
     * otherwise it'll return basic information about this doc.
     */
    get doc() {
        const channel = this.client.getChannel(this.raw.serverId, this.raw.reaction.channelId);
        return channel?.docs?.get(this.docID) ?? {
            id: this.docID,
            guild: this.client.guilds.get(this.raw.serverId) ?? {
                id: this.raw.serverId
            },
            channelID: this.raw.reaction.channelId
        };
    }
}
exports.DocReactionInfo = DocReactionInfo;
