// touchguild/dist/lib/structures/ForumThread.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ForumThread = void 0;
const tslib_1 = require("tslib");
const Base_1 = require("./Base");
const ForumThreadComment_1 = require("./ForumThreadComment");
const TypedCollection_1 = tslib_1.__importDefault(require("../util/TypedCollection"));
/** Represents a thread/topic coming from a "Forums" channel. */
class ForumThread extends Base_1.Base {
    /**
     * @param data raw data
     * @param client client
     */
    constructor(data, client) {
        super(data.id, client);
        this.guildID = data.serverId;
        this.channelID = data.channelId;
        this.name = data.title;
        this.createdAt = new Date(data.createdAt);
        this.ownerID = data.createdBy;
        this.owner = (this.client.getMember(data.serverId, data.createdBy) ?? this.client.users.get(data.createdBy) ?? this.client.rest.guilds.getMember(data.serverId, data.createdBy));
        this.editedTimestamp = data.updatedAt ? new Date(data.updatedAt) : null;
        this.bumpedAt = data.bumpedAt ? new Date(data.bumpedAt) : null;
        this.content = data.content;
        this.mentions = data.mentions ?? null;
        this.comments = new TypedCollection_1.default(ForumThreadComment_1.ForumThreadComment, client, client.params.collectionLimits?.threadComments);
        this.isLocked = data.isLocked ?? false;
        this.isPinned = data.isPinned ?? false;
        this.update(data);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            guildID: this.guildID,
            channelID: this.channelID,
            name: this.name,
            createdAt: this.createdAt,
            owner: this.owner,
            ownerID: this.ownerID,
            editedTimestamp: this.editedTimestamp,
            bumpedAt: this.bumpedAt,
            content: this.content,
            mentions: this.mentions,
            comments: this.comments.map(comment => comment.toJSON()),
            isLocked: this.isLocked,
            isPinned: this.isPinned
        };
    }
    update(data) {
        if (data.bumpedAt !== undefined) {
            this.bumpedAt = new Date(data.bumpedAt);
        }
        if (data.channelId !== undefined) {
            this.channelID = data.channelId;
        }
        if (data.content !== undefined) {
            this.content = data.content;
        }
        if (data.createdAt !== undefined) {
            this.createdAt = new Date(data.createdAt);
        }
        if (data.createdBy !== undefined) {
            this.ownerID = data.createdBy;
        }
        if (data.id !== undefined) {
            this.id = data.id;
        }
        if (data.isLocked !== undefined) {
            this.isLocked = data.isLocked;
        }
        if (data.mentions !== undefined) {
            this.mentions = data.mentions ?? null;
        }
        if (data.serverId !== undefined) {
            this.guildID = data.serverId;
        }
        if (data.title !== undefined) {
            this.name = data.title;
        }
        if (data.updatedAt !== undefined) {
            this.editedTimestamp = new Date(data.updatedAt);
        }
    }
    /** The guild the thread is in. This will throw an error if the guild isn't cached.*/
    get guild() {
        if (!this._cachedGuild) {
            this._cachedGuild = this.client.getGuild(this.guildID);
            if (!this._cachedGuild) {
                throw new Error(`${this.constructor.name}#guild: couldn't find the Guild in cache.`);
            }
        }
        return this._cachedGuild;
    }
    /** The forum channel this thread was created in.  */
    get channel() {
        return this._cachedChannel ?? (this._cachedChannel = this.client.getChannel(this.guildID, this.channelID));
    }
    /** Add a comment to this forum thread.
     * @param options Options of the comment.
     */
    async createForumComment(options) {
        return this.client.rest.channels.createForumComment(this.channelID, this.id, options);
    }
    /** Edit the forum thread.
     * @param options Edit options.
     */
    async edit(options) {
        return this.client.rest.channels.editForumThread(this.channelID, this.id, options);
    }
    /** Delete this forum thread. */
    async delete() {
        return this.client.rest.channels.deleteForumThread(this.channelID, this.id);
    }
    /** Pin this forum thread. */
    async pin() {
        return this.client.rest.channels.pinForumThread(this.channelID, this.id);
    }
    /** Unpin this forum thread. */
    async unpin() {
        return this.client.rest.channels.unpinForumThread(this.channelID, this.id);
    }
    /** Lock this forum thread. */
    async lock() {
        return this.client.rest.channels.lockForumThread(this.channelID, this.id);
    }
    /** Unlock this forum thread. */
    async unlock() {
        return this.client.rest.channels.unlockForumThread(this.channelID, this.id);
    }
    /** Add a reaction to this forum thread.
     * @param emoteID ID of the emote to be added.
     */
    async createReaction(emoteID) {
        return this.client.rest.channels.createReaction(this.channelID, "ForumThread", this.id, emoteID);
    }
    /** Remove a reaction from this forum thread.
     * @param emoteID ID of the emote to be added.
     */
    async deleteReaction(emoteID) {
        return this.client.rest.channels.deleteReaction(this.channelID, "ForumThread", this.id, emoteID);
    }
}
exports.ForumThread = ForumThread;
