// touchguild/dist/lib/structures/ForumChannel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ForumChannel = void 0;
const tslib_1 = require("tslib");
const ForumThread_1 = require("./ForumThread");
const GuildChannel_1 = require("./GuildChannel");
const TypedCollection_1 = tslib_1.__importDefault(require("../util/TypedCollection"));
/** Represents a forum channel. */
class ForumChannel extends GuildChannel_1.GuildChannel {
    /**
     * @param data raw data
     * @param client client
     */
    constructor(data, client) {
        super(data, client);
        this.threads = new TypedCollection_1.default(ForumThread_1.ForumThread, client, client.params.collectionLimits?.threads);
        this.update(data);
    }
    /** Create a thread in this channel.
     * @param options Thread's options including title & content.
     */
    async createThread(options) {
        return this.client.rest.channels.createForumThread(this.id, options);
    }
    /** Edit a thread from this channel.
     * @param threadID ID of a thread.
     * @param options Edit options.
     */
    async editThread(threadID, options) {
        return this.client.rest.channels.editForumThread(this.id, threadID, options);
    }
    /** Delete a thread from this channel.
     * @param threadID ID of a thread.
     */
    async deleteThread(threadID) {
        return this.client.rest.channels.deleteForumThread(this.id, threadID);
    }
    /** Pin a thread.
     * @param threadID ID of a thread.
     */
    async pinThread(threadID) {
        return this.client.rest.channels.pinForumThread(this.id, threadID);
    }
    /** Unpin a thread.
     * @param threadID ID of a thread.
     */
    async unpinThread(threadID) {
        return this.client.rest.channels.unpinForumThread(this.id, threadID);
    }
    /** Lock a thread.
     * @param threadID ID of a thread.
     */
    async lockThread(threadID) {
        return this.client.rest.channels.lockForumThread(this.id, threadID);
    }
    /** Unlock a thread.
     * @param threadID ID of a thread.
     */
    async unlockThread(threadID) {
        return this.client.rest.channels.unlockForumThread(this.id, threadID);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            threads: this.threads.map(thread => thread.toJSON())
        };
    }
}
exports.ForumChannel = ForumChannel;
