// touchguild/dist/lib/structures/ForumThreadComment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ForumThreadComment = void 0;
const Base_1 = require("./Base");
/** Represents a comment coming from a ForumThread. */
class ForumThreadComment extends Base_1.Base {
    constructor(data, client, options) {
        super(data.id, client);
        this.content = data.content;
        this.createdAt = new Date(data.createdAt);
        this.updatedAt = data.updatedAt ? new Date(data.updatedAt) : null;
        this.channelID = data.channelId;
        this.threadID = data.forumTopicId;
        this.memberID = data.createdBy;
        this.guildID = options?.guildID ?? null;
        this.mentions = data.mentions ?? null;
        this.update(data);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            content: this.content,
            createdAt: this.createdAt,
            updatedAt: this.updatedAt,
            channelID: this.channelID,
            threadID: this.threadID,
            memberID: this.memberID,
            guildID: this.guildID,
            mentions: this.mentions
        };
    }
    update(data) {
        if (data.channelId !== undefined) {
            this.channelID = data.channelId;
        }
        if (data.content !== undefined) {
            this.content = data.content;
        }
        if (data.createdAt !== undefined) {
            this.createdAt = new Date(data.createdAt);
        }
        if (data.createdBy !== undefined) {
            this.memberID = data.createdBy;
        }
        if (data.forumTopicId !== undefined) {
            this.threadID = data.forumTopicId;
        }
        if (data.id !== undefined) {
            this.id = data.id;
        }
        if (data.mentions !== undefined) {
            this.mentions = data.mentions;
        }
        if (data.updatedAt !== undefined) {
            this.updatedAt = new Date(data.updatedAt);
        }
        this.guildID = this.toJSON().guildID;
    }
    /** Retrieve the member who sent this comment, if cached.
     * If there is no cached member, this will make a rest request which returns a Promise.
     * If the request fails, it'll return you undefined as a value.
     */
    get member() {
        return (this.client.getGuild(this.guildID)?.members.get(this.memberID) ?? this.guildID) ? this.client.rest.guilds.getMember(this.guildID, this.memberID) : undefined;
    }
    /** Add a comment to the same forum thread as this comment.
     * @param options New comment's options.
     */
    async createForumComment(options) {
        return this.client.rest.channels.createForumComment(this.channelID, this.threadID, options);
    }
    /** Add a reaction to the comment.
     * @param reaction The ID of the reaction to add.
     */
    async createReaction(reaction) {
        return this.client.rest.channels.createReactionToSubcategory(this.channelID, "ForumThreadComment", this.threadID, this.id, reaction);
    }
    /** Remove a reaction from the comment.
     * @param reaction The ID of the reaction to remove.
     */
    async deleteReaction(reaction) {
        return this.client.rest.channels.deleteReactionFromSubcategory(this.channelID, "ForumThreadComment", this.threadID, this.id, reaction);
    }
    /** Edit this forum thread's comment.
     * @param options Edit options.
     */
    async edit(options) {
        return this.client.rest.channels.editForumComment(this.channelID, this.threadID, this.id, { content: options?.content });
    }
    /** Delete this forum thread comment. */
    async delete() {
        return this.client.rest.channels.deleteForumComment(this.channelID, this.threadID, this.id);
    }
}
exports.ForumThreadComment = ForumThreadComment;
