// touchguild/dist/lib/structures/ForumThreadReactionInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ForumThreadReactionInfo = void 0;
/** @module ForumThreadReactionInfo */
const ReactionInfo_1 = require("./ReactionInfo");
/** Information about a ForumThread's reaction. */
class ForumThreadReactionInfo extends ReactionInfo_1.ReactionInfo {
    /**
     * @param data raw data.
     * @param client client.
     */
    constructor(data, client) {
        super(data, client);
        this.threadID = data.reaction.forumTopicId;
        this.commentID = data.reaction["forumTopicCommentId"] ?? null;
        this.type = data.reaction["forumTopicCommentId"] ? "comment" : "thread";
    }
    /** The forum thread where the reaction has been added.
     * If the thread is cached, it'll return a ForumThread component,
     * otherwise it'll return basic information about this thread.
     */
    get thread() {
        return this.client.getChannel(this.raw.serverId, this.raw.reaction.channelId)?.threads.get(this.threadID) ?? {
            id: this.threadID,
            guild: this.client.guilds.get(this.raw.serverId) ?? {
                id: this.raw.serverId
            },
            channelID: this.raw.reaction.channelId
        };
    }
}
exports.ForumThreadReactionInfo = ForumThreadReactionInfo;
