// touchguild/dist/lib/structures/Guild.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Guild = void 0;
const tslib_1 = require("tslib");
const Base_1 = require("./Base");
const Member_1 = require("./Member");
const GuildChannel_1 = require("./GuildChannel");
const GuildGroup_1 = require("./GuildGroup");
const GuildRole_1 = require("./GuildRole");
const TypedCollection_1 = tslib_1.__importDefault(require("../util/TypedCollection"));
/** Represents a Guild, also called server. */
class Guild extends Base_1.Base {
    /**
     * @param data raw data.
     * @param client client.
     */
    constructor(data, client) {
        super(data.id, client);
        this.ownerID = data.ownerId;
        this.type = data.type;
        this.name = data.name;
        this.url = data.url;
        this.description = data.about;
        this.iconURL = data.avatar ?? null;
        this.bannerURL = data.banner ?? null;
        this.timezone = data.timezone;
        this.defaultChannelID = data.defaultChannelId;
        this.createdAt = new Date(data.createdAt);
        this.verified = data.isVerified ?? false;
        this.groups = new TypedCollection_1.default(GuildGroup_1.GuildGroup, client);
        this.channels = new TypedCollection_1.default(GuildChannel_1.GuildChannel, client);
        this.members = new TypedCollection_1.default(Member_1.Member, client);
        this.roles = new TypedCollection_1.default(GuildRole_1.GuildRole, client);
        this.update(data);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            ownerID: this.ownerID,
            type: this.type,
            name: this.name,
            url: this.url,
            description: this.description,
            iconURL: this.iconURL,
            bannerURL: this.bannerURL,
            timezone: this.timezone,
            defaultChannelID: this.defaultChannelID,
            createdAt: this.createdAt,
            verified: this.verified,
            channels: this.channels.map(channel => channel.toJSON()),
            members: this.members.map(member => member.toJSON())
        };
    }
    update(data) {
        if (data.about !== undefined) {
            this.description = data.about;
        }
        if (data.avatar !== undefined) {
            this.iconURL = data.avatar;
        }
        if (data.banner !== undefined) {
            this.bannerURL = data.banner;
        }
        if (data.createdAt !== undefined) {
            this.createdAt = new Date(data.createdAt);
        }
        if (data.defaultChannelId !== undefined) {
            this.defaultChannelID = data.defaultChannelId;
        }
        if (data.id !== undefined) {
            this.id = data.id;
        }
        if (data.isVerified !== undefined) {
            this.verified = data.isVerified;
        }
        if (data.name !== undefined) {
            this.name = data.name;
        }
        if (data.ownerId !== undefined) {
            this.ownerID = data.ownerId;
        }
        if (data.timezone !== undefined) {
            this.timezone = data.timezone;
        }
        if (data.type !== undefined) {
            this.type = data.type;
        }
        if (data.url !== undefined) {
            this.url = data.url;
        }
    }
    /** Retrieve cached or rest guild's owner. */
    get owner() {
        return this.client.getGuild(this.id)?.members.get(this.ownerID) ?? this.client.users.get(this.ownerID) ?? this.client.rest.guilds.getMember(this.id, this.ownerID);
    }
    /** Get a channel from this guild, if cached.
     * @param channelID The ID of the channel to get from cache.
     */
    getChannel(channelID) {
        if (!channelID)
            throw new Error("channelID is a required parameter.");
        return this.channels.get(channelID);
    }
    /** Get a member from this guild, if cached.
     * @param memberID The ID of the member to get.
     */
    getMember(memberID) {
        if (!memberID)
            throw new Error("memberID is a required parameter.");
        return this.members.get(memberID);
    }
    /** Ban a member.
     * @param memberID ID of the member to ban.
     * @param reason The reason of the ban.
     */
    async createBan(memberID, reason) {
        return this.client.rest.guilds.createBan(this.id, memberID, reason);
    }
    /** Unban a member.
     * @param memberID ID of the member to unban.
     */
    async removeBan(memberID) {
        return this.client.rest.guilds.removeBan(this.id, memberID);
    }
    /** Get Subscription
     * @param subscriptionID ID of the subscription to get.
     */
    async getSubscription(subscriptionID) {
        return this.client.rest.guilds.getSubscription(this.id, subscriptionID);
    }
    /** Get Subscriptions */
    async getSubscriptions() {
        return this.client.rest.guilds.getSubscriptions(this.id);
    }
    /** Bulk Award XP Members
     * @param options Members to award XP and amount of XP to award.
     */
    async bulkAwardXPMembers(options) {
        return this.client.rest.guilds.bulkAwardXP(this.id, options);
    }
    /** Bulk set XP Members
     * @param options Members to set XP and amount of XP to set.
     */
    async bulkSetXPMembers(options) {
        return this.client.rest.guilds.bulkSetXP(this.id, options);
    }
    /**
     * Create a category
     * @param guildID ID of the guild.
     * @param options Create options.
     */
    async createCategory(options) {
        return this.client.rest.guilds.createCategory(this.id, options);
    }
    /**
     * Read a guild category.
     * @param guildID ID of the guild to create a category in.
     * @param categoryID ID of the category you want to read.
     */
    async getCategory(categoryID) {
        return this.client.rest.guilds.getCategory(this.id, categoryID);
    }
    /**
     * Edit a category.
     * @param guildID ID of the guild to create a category in.
     * @param categoryID ID of the category you want to read.
     * @param options Options to update a category.
     */
    async editCategory(categoryID, options) {
        return this.client.rest.guilds.editCategory(this.id, categoryID, options);
    }
    /**
     * Delete a category.
     * @param guildID ID of the guild to create a category in.
     * @param categoryID ID of the category you want to read.
     */
    async deleteCategory(categoryID) {
        return this.client.rest.guilds.deleteCategory(this.id, categoryID);
    }
}
exports.Guild = Guild;
