// touchguild/dist/lib/structures/GuildCategory.js
"use strict";
/** @module GuildCategory */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GuildCategory = void 0;
const Base_1 = require("./Base");
/** Class representing a guild group. */
class GuildCategory extends Base_1.Base {
    constructor(data, client) {
        super(data.id, client);
        this.id = data.id;
        this.guildID = data.serverId;
        this.groupID = data.groupId;
        this.createdAt = new Date(data.createdAt);
        this.updatedAt = data.updatedAt ?? null;
        this.name = data.name;
        this.update(data);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            guildID: this.guildID,
            groupID: this.groupID,
            createdAt: this.createdAt,
            updatedAt: this.updatedAt ?? null,
            name: this.name
        };
    }
    update(data) {
        if (data.id !== undefined) {
            this.id = data.id;
        }
        if (data.serverId !== undefined) {
            this.guildID = data.serverId;
        }
        if (data.groupId !== undefined) {
            this.groupID = data.groupId;
        }
        if (data.createdAt !== undefined) {
            this.createdAt = new Date(data.createdAt);
        }
        if (data.updatedAt !== undefined) {
            this.updatedAt = data.updatedAt ?? null;
        }
        if (data.name !== undefined) {
            this.name = data.name;
        }
    }
    /**
     * Update a guild category.
     * @param guildID ID of the guild to create a category in.
     * @param categoryID ID of the category you want to read.
     * @param options Options to update a category.
     */
    async editCategory(options) {
        return this.client.rest.guilds.editCategory(this.guildID, this.id, options);
    }
    /**
     * Delete a guild category.
     * @param guildID ID of the guild to create a category in.
     * @param categoryID ID of the category you want to read.
     */
    async deleteCategory() {
        return this.client.rest.guilds.deleteCategory(this.guildID, this.id);
    }
    /**
     * Create a channel category permission assigned to a user or role.
     * @param targetID ID of the user (string) or role (number) to assign the permission to
     * @param options Permission options
     *
     * Warning: targetID must have the correct type (number=role, string=user).
     */
    async createPermission(targetID, options) {
        return this.client.rest.guilds.createCategoryPermission(this.guildID, this.id, targetID, options);
    }
    /**
     * Update a category permission.
     *
     * Warning: targetID must have the correct type (number=role, string=user).
     */
    async editPermission(targetID, options) {
        return this.client.rest.guilds.editCategoryPermission(this.guildID, this.id, targetID, options);
    }
    /**
     * Get permission coming from a category.
     * @param targetID ID of the user (string) or role (number) to get the permission for
     *
     * Warning: targetID must have the correct type (number=role, string=user).
     */
    async getPermission(targetID) {
        return this.client.rest.guilds.getCategoryPermission(this.guildID, this.id, targetID);
    }
    /**
     * Get permissions of a category.
     */
    async getPermissions() {
        return this.client.rest.guilds.getCategoryPermissions(this.guildID, this.id);
    }
    /**
     * Get user permissions from a specified category.
     */
    async getUserPermissions() {
        return this.client.rest.guilds.getCategoryUserPermissions(this.guildID, this.id);
    }
    /**
     * Get role permissions from a specified category.
     */
    async getRolePermissions() {
        return this.client.rest.guilds.getCategoryRolePermissions(this.guildID, this.id);
    }
    /**
     * Delete a category permission.
     * @param targetID ID of the user or role to delete the permission from
     */
    async deletePermission(targetID) {
        return this.client.rest.guilds.deleteCategoryPermission(this.guildID, this.id, targetID);
    }
}
exports.GuildCategory = GuildCategory;
