// touchguild/dist/lib/structures/GuildRole.js
"use strict";
/** @module GuildRole */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GuildRole = void 0;
const Base_1 = require("./Base");
/** Class representing a guild role. */
class GuildRole extends Base_1.Base {
    constructor(data, client) {
        super(data.id, client);
        this.guildID = data.serverId;
        this.createdAt = new Date(data.createdAt);
        this.editedTimestamp = data.updatedAt ? new Date(data.updatedAt) : null;
        this.name = data.name;
        this.isDisplayedSeparately = data.isDisplayedSeparately ?? false;
        this.isSelfAssignable = data.isSelfAssignable ?? false;
        this.isMentionable = data.isMentionable ?? false;
        this.permissions = data.permissions;
        this.colors = data.colors ?? null;
        this.iconURL = data.icon ?? null;
        this.position = data.priority;
        this.isBase = data.isBase ?? false;
        this.botUserID = data.botUserId ?? null;
        this.update(data);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            guildID: this.guildID,
            createdAt: this.createdAt,
            editedTimestamp: this.editedTimestamp,
            name: this.name,
            isDisplayedSeparately: this.isDisplayedSeparately,
            isSelfAssignable: this.isSelfAssignable,
            isMentionable: this.isMentionable,
            permissions: this.permissions,
            colors: this.colors,
            iconURL: this.iconURL,
            position: this.position,
            isBase: this.isBase,
            botUserID: this.botUserID
        };
    }
    update(data) {
        if (data.serverId !== undefined) {
            this.guildID = data.serverId;
        }
        if (data.createdAt !== undefined) {
            this.createdAt = new Date(data.createdAt);
        }
        if (data.updatedAt !== undefined) {
            this.editedTimestamp = new Date(data.updatedAt);
        }
        if (data.name !== undefined) {
            this.name = data.name;
        }
        if (data.isDisplayedSeparately !== undefined) {
            this.isDisplayedSeparately = data.isDisplayedSeparately;
        }
        if (data.isSelfAssignable !== undefined) {
            this.isSelfAssignable = data.isSelfAssignable;
        }
        if (data.isMentionable !== undefined) {
            this.isMentionable = data.isMentionable;
        }
        if (data.permissions !== undefined) {
            this.permissions = data.permissions;
        }
        if (data.colors !== undefined) {
            this.colors = data.colors;
        }
        if (data.icon !== undefined) {
            this.iconURL = data.icon;
        }
        if (data.priority !== undefined) {
            this.position = data.priority;
        }
        if (data.isBase !== undefined) {
            this.isBase = data.isBase;
        }
    }
    /** Edit the role permission */
    async editPermission(options) {
        return this.client.rest.guilds.editRolePermission(this.guildID, this.id, options);
    }
}
exports.GuildRole = GuildRole;
