// touchguild/dist/lib/structures/GuildSubscription.js
"use strict";
/** @module GuildSubscription */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GuildSubscription = void 0;
const Base_1 = require("./Base");
/** Class representing a guild group. */
class GuildSubscription extends Base_1.Base {
    constructor(data, client) {
        super(data.serverId, client);
        this.type = data.type;
        this.guildID = data.serverId;
        this.description = data.description !== "" ? data.description : null;
        this.roleID = data.roleId ?? null;
        this.cost = data.cost;
        this.createdAt = new Date(data.createdAt);
        this.update(data);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            type: this.type,
            guildID: this.guildID,
            description: this.description,
            roleID: this.roleID,
            cost: this.cost,
            createdAt: this.createdAt
        };
    }
    update(data) {
        if (data.serverId !== undefined) {
            this.guildID = data.serverId;
        }
        if (data.type !== undefined) {
            this.type = data.type;
        }
        if (data.description !== undefined) {
            this.description = data.description !== "" ? data.description : null;
        }
        if (data.roleId !== undefined) {
            this.roleID = data.roleId;
        }
        if (data.cost !== undefined) {
            this.cost = data.cost;
        }
        if (data.createdAt !== undefined) {
            this.createdAt = new Date(data.createdAt);
        }
    }
}
exports.GuildSubscription = GuildSubscription;
