// touchguild/dist/lib/structures/GuildChannel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GuildChannel = void 0;
const Base_1 = require("./Base");
/** Represents a guild channel. */
class GuildChannel extends Base_1.Base {
    // /** Cached messages. */
    // messages: TypedCollection<string, APIChatMessage, Message<AnyTextableChannel>>;
    // /** Cached threads. */
    // threads: TypedCollection<number, APIForumTopic, ForumThread<AnyTextableChannel>>;
    // /** Cached docs. */
    // docs: TypedCollection<number, APIDoc, Doc>;
    // /** Cached scheduled events. */
    // scheduledEvents: TypedCollection<number, APICalendarEvent, CalendarEvent>;
    /**
     * @param data raw data
     * @param client client
     */
    constructor(data, client) {
        super(data.id, client);
        this.type = data.type;
        this.name = data.name;
        this.description = data.topic ?? null;
        this.createdAt = new Date(data.createdAt);
        this.creatorID = data.createdBy;
        this.editedTimestamp = data.updatedAt ? new Date(data.updatedAt) : null;
        this.guildID = data.serverId;
        this.parentID = data.parentId ?? null;
        this.categoryID = data.categoryId ?? null;
        this.groupID = data.groupId;
        this.archivedBy = data.archivedBy ?? null;
        this.archivedAt = data.archivedAt ? new Date(data.archivedAt) : null;
        this.visibility = data.visibility ?? "public";
        this.isPublic = this.visibility === "public" ? true : false;
        // this.messages = new TypedCollection(Message, client, client.params.collectionLimits?.messages);
        // this.threads = new TypedCollection(ForumThread, client, client.params.collectionLimits?.threads);
        // this.docs = new TypedCollection(Doc, client, client.params.collectionLimits?.docs);
        // this.scheduledEvents = new TypedCollection(CalendarEvent, client, client.params.collectionLimits?.scheduledEvents);
        this.update(data);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            type: this.type,
            name: this.name,
            description: this.description,
            createdAt: this.createdAt,
            creatorID: this.creatorID,
            editedTimestamp: this.editedTimestamp,
            guildID: this.guildID,
            parentID: this.parentID,
            categoryID: this.categoryID,
            groupID: this.groupID,
            isPublic: this.isPublic,
            archivedBy: this.archivedBy,
            archivedAt: this.archivedAt,
            visibility: this.visibility
        };
    }
    update(data) {
        if (data.archivedAt !== undefined) {
            this.archivedAt = new Date(data.archivedAt);
        }
        if (data.archivedBy !== undefined) {
            this.archivedBy = data.archivedBy;
        }
        if (data.categoryId !== undefined) {
            this.categoryID = data.categoryId;
        }
        if (data.createdAt !== undefined) {
            this.createdAt = new Date(data.createdAt);
        }
        if (data.createdBy !== undefined) {
            this.creatorID = data.createdBy;
        }
        if (data.groupId !== undefined) {
            this.groupID = data.groupId;
        }
        if (data.id !== undefined) {
            this.id = data.id;
        }
        if (data.isPublic !== undefined) {
            this.isPublic = data.isPublic;
        }
        if (data.name !== undefined) {
            this.name = data.name;
        }
        if (data.parentId !== undefined) {
            this.parentID = data.parentId;
        }
        if (data.serverId !== undefined) {
            this.guildID = data.serverId;
        }
        if (data.topic !== undefined) {
            this.description = data.topic;
        }
        if (data.type !== undefined) {
            this.type = data.type;
        }
        if (data.updatedAt !== undefined) {
            this.editedTimestamp = new Date(data.updatedAt);
        }
        if (data.visibility !== undefined) {
            this.visibility = data.visibility;
        }
    }
    /** Edit the channel. */
    async edit(options) {
        return this.client.rest.guilds.editChannel(this.id, options);
    }
    /** Delete the channel. */
    async delete() {
        return this.client.rest.guilds.deleteChannel(this.id);
    }
    /** Archive the channel */
    async archive() {
        return this.client.rest.channels.archiveChannel(this.id);
    }
    /** Unarchive the channel */
    async restore() {
        return this.client.rest.channels.restoreChannel(this.id);
    }
}
exports.GuildChannel = GuildChannel;
