// touchguild/dist/lib/structures/GuildGroup.js
"use strict";
/** @module GuildGroup */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GuildGroup = void 0;
const Base_1 = require("./Base");
/** Class representing a guild group. */
class GuildGroup extends Base_1.Base {
    constructor(data, client) {
        super(data.id, client);
        this.guildID = data.serverId;
        this.name = data.name;
        this.description = data.description ?? null;
        this.avatarURL = data.avatar ?? null;
        this.isHome = data.isHome ?? false;
        this.emoteID = data.emoteId ?? null;
        this.isPublic = data.isPublic ?? false;
        this.createdAt = new Date(data.createdAt);
        this.createdBy = data.createdBy;
        this.editedTimestamp = data.updatedAt ? new Date(data.updatedAt) : null;
        this.updatedBy = data.updatedBy ?? null;
        this.archivedAt = data.archivedAt ? new Date(data.archivedAt) : null;
        this.archivedBy = data.archivedBy ?? null;
        this.update(data);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            guildID: this.guildID,
            name: this.name,
            description: this.description,
            avatarURL: this.avatarURL,
            isHome: this.isHome,
            emoteID: this.emoteID,
            isPublic: this.isPublic,
            createdAt: this.createdAt,
            createdBy: this.createdBy,
            editedTimestamp: this.editedTimestamp,
            updatedBy: this.updatedBy,
            archivedAt: this.archivedAt,
            archivedBy: this.archivedBy
        };
    }
    update(data) {
        if (data.serverId !== undefined) {
            this.guildID = data.serverId;
        }
        if (data.name !== undefined) {
            this.name = data.name;
        }
        if (data.description !== undefined) {
            this.description = data.description;
        }
        if (data.avatar !== undefined) {
            this.avatarURL = data.avatar;
        }
        if (data.isHome !== undefined) {
            this.isHome = data.isHome;
        }
        if (data.emoteId !== undefined) {
            this.emoteID = data.emoteId;
        }
        if (data.isPublic !== undefined) {
            this.isPublic = data.isPublic;
        }
        if (data.createdAt !== undefined) {
            this.createdAt = new Date(data.createdAt);
        }
        if (data.createdBy !== undefined) {
            this.createdBy = data.createdBy;
        }
        if (data.updatedAt !== undefined) {
            this.editedTimestamp = new Date(data.updatedAt);
        }
        if (data.updatedBy !== undefined) {
            this.updatedBy = data.updatedBy;
        }
        if (data.archivedAt !== undefined) {
            this.archivedAt = new Date(data.archivedAt);
        }
        if (data.archivedBy !== undefined) {
            this.archivedBy = data.archivedBy;
        }
    }
}
exports.GuildGroup = GuildGroup;
