// touchguild/dist/lib/structures/ListItem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListItem = void 0;
const Base_1 = require("./Base");
/** Represents an item of a "Lists" channel. */
class ListItem extends Base_1.Base {
    /**
     * @param data raw data.
     * @param client client.
     */
    constructor(data, client) {
        super(data.id, client);
        this._data = data;
        this.guildID = data.serverId;
        this.channelID = data.channelId;
        this.content = data.message ?? null;
        this.mentions = data.mentions ?? null;
        this.createdAt = data.createdAt ? new Date(data.createdAt) : null;
        this.memberID = data.createdBy;
        this.webhookID = data.createdByWebhookId ?? null;
        this.editedTimestamp = data.updatedAt ? new Date(data.updatedAt) : null;
        this.updatedBy = data.updatedBy ?? null;
        this.parentListItemID = data.parentListItemId ?? null;
        this.completedAt = data.completedAt ? new Date(data.completedAt) : null;
        this.completedBy = data.completedBy ?? null;
        this.update(data);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            guildID: this.guildID,
            channelID: this.channelID,
            content: this.content,
            mentions: this.mentions,
            createdAt: this.createdAt,
            memberID: this.memberID,
            webhookID: this.webhookID,
            editedTimestamp: this.editedTimestamp,
            updatedBy: this.updatedBy,
            parentListItemID: this.parentListItemID,
            completedAt: this.completedAt,
            completedBy: this.completedBy
        };
    }
    update(data) {
        if (data.channelId !== undefined) {
            this.channelID = data.channelId;
        }
        if (data.completedAt !== undefined) {
            this.completedAt = new Date(data.completedAt);
        }
        if (data.completedBy !== undefined) {
            this.completedBy = data.completedBy;
        }
        if (data.createdAt !== undefined) {
            this.createdAt = new Date(data.createdAt);
        }
        if (data.createdBy !== undefined) {
            this.memberID = data.createdBy;
        }
        if (data.createdByWebhookId !== undefined) {
            this.webhookID = data.createdByWebhookId;
        }
        if (data.id !== undefined) {
            this.id = data.id;
        }
        if (data.mentions !== undefined) {
            this.mentions = data.mentions;
        }
        if (data.message !== undefined) {
            this.content = data.message;
        }
        if (data.note !== undefined) {
            this._data.note = data.note;
        }
        if (data.parentListItemId !== undefined) {
            this.parentListItemID = data.parentListItemId;
        }
        if (data.serverId !== undefined) {
            this.guildID = data.serverId;
        }
        if (data.updatedAt !== undefined) {
            this.editedTimestamp = new Date(data.updatedAt);
        }
        if (data.updatedBy !== undefined) {
            this.updatedBy = data.updatedBy;
        }
    }
    get note() {
        return this._data.note ? {
            createdAt: new Date(this._data.note.createdAt),
            memberID: this._data.note.createdBy,
            editedTimestamp: this._data.note.updatedAt ? new Date(this._data.note.updatedAt) : null,
            editedBy: this._data.note.updatedBy ?? null,
            mentions: this._data.note.mentions ?? null,
            content: this._data.note.content
        } : null;
    }
    /** Retrieve the member who executed this action.
     *
     * Note: If the item has been edited, the updatedBy id will be used to get you the member.
     */
    get member() {
        return this.client.getGuild(this.guildID)?.members.get(this.updatedBy ?? this.memberID) ?? this.client.rest.guilds.getMember(this.guildID, this.updatedBy ?? this.memberID);
    }
    /** Edit this item.
     * @param options Edit options.
     */
    async edit(options) {
        return this.client.rest.channels.editListItem(this.channelID, this.id, options);
    }
    /** Delete this item. */
    async delete() {
        return this.client.rest.channels.deleteListItem(this.channelID, this.id);
    }
    /** Set this item as "complete". */
    async complete() {
        return this.client.rest.channels.completeListItem(this.channelID, this.id);
    }
    /** Set this item as "uncomplete". */
    async uncomplete() {
        return this.client.rest.channels.uncompleteListItem(this.channelID, this.id);
    }
}
exports.ListItem = ListItem;
