// touchguild/dist/lib/structures/Member.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Member = void 0;
const User_1 = require("./User");
/** Represents a guild user. */
class Member extends User_1.User {
    /**
     * @param data raw data.
     * @param client client.
     * @param guildID ID of the guild.
     */
    constructor(data, client, guildID) {
        super(data.user, client);
        this._data = data;
        this.roles = data.roleIds ?? null;
        this.nickname = data.nickname ?? null;
        this.joinedAt = data.joinedAt ? new Date(data.joinedAt) : null;
        this.isOwner = data.isOwner ?? false;
        this.guildID = guildID;
        this.update(data);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            roles: this.roles,
            nickname: this.nickname,
            joinedAt: this.joinedAt,
            isOwner: this.isOwner,
            guildID: this.guildID
        };
    }
    update(data) {
        if (data.isOwner !== undefined) {
            this.isOwner = data.isOwner ?? false;
        }
        if (data.joinedAt !== undefined) {
            this.joinedAt = data.joinedAt ? new Date(data.joinedAt) : null;
        }
        if (data.nickname !== undefined) {
            this.nickname = data.nickname ?? null;
        }
        if (data.roleIds !== undefined) {
            this.roles = data.roleIds ?? null;
        }
        if (data.user !== undefined) {
            super.update(data.user);
            this.client.users.update(data.user);
        }
    }
    /** Guild where the user comes from, returns Guild or a promise.
     * If guild isn't cached & the request failed, this will return you undefined.
     */
    get guild() {
        return this.client.guilds.get(this.guildID) ?? this.client.rest.guilds.getGuild(this.guildID);
    }
    /** Member's user, shows less information. */
    get user() {
        if (this.client.users.get(this.id)) {
            return this.client.users.update(this._data.user);
        }
        else {
            const USER = new User_1.User(this._data.user, this.client);
            this.client.users.add(USER);
            return USER;
        }
    }
    /** Edit this member.
     * @param options Edit options.
     */
    async edit(options) {
        return this.client.rest.guilds.editMember(this.guildID, this.id, options);
    }
    /** Kick this member. */
    async kick() {
        return this.client.rest.guilds.removeMember(this.guildID, this.id);
    }
    /** Ban this member.
     * @param reason The reason of the ban.
     */
    async ban(reason) {
        return this.client.rest.guilds.createBan(this.guildID, this.id, reason);
    }
    /** Unban this member. */
    async unban() {
        return this.client.rest.guilds.removeBan(this.guildID, this.id);
    }
    /** Get a specified social link from the member, if member is connected to them through Guilded.
     * @param socialMediaName Name of a social media linked to this member.
     */
    async getSocialLink(socialMediaName) {
        return this.client.rest.misc.getSocialLink(this.guildID, this.id, socialMediaName);
    }
    /** Add this member to a guild group.
     * @param groupID ID of the guild group.
     */
    async addGroup(groupID) {
        return this.client.rest.guilds.memberAddGroup(groupID, this.id);
    }
    /** Remove this member from a guild group.
     * @param groupID ID of the guild group.
     */
    async removeGroup(groupID) {
        return this.client.rest.guilds.memberRemoveGroup(groupID, this.id);
    }
    /** Add a role to this member.
     * @param roleID ID of the role to be added.
     */
    async addRole(roleID) {
        return this.client.rest.guilds.memberAddRole(this.guildID, this.id, roleID);
    }
    /** Remove a role from this member.
     * @param roleID ID of the role to be added.
     */
    async removeRole(roleID) {
        return this.client.rest.guilds.memberRemoveRole(this.guildID, this.id, roleID);
    }
    /** Award the member using the built-in EXP system.
     * @param amount Amount of experience to give.
     */
    async award(amount) {
        return this.client.rest.guilds.awardMember(this.guildID, this.id, amount);
    }
    /** Set member's experience using the built-in EXP system.
     * @param amount Amount of experience to set.
     */
    async setXP(amount) {
        return this.client.rest.guilds.setMemberXP(this.guildID, this.id, amount);
    }
    /** Get member permission */
    async getPermission() {
        return this.client.rest.guilds.getMemberPermission(this.guildID, this.id);
    }
}
exports.Member = Member;
