// touchguild/dist/lib/structures/MessageReactionInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageReactionInfo = void 0;
/** @module MessageReactionInfo */
const ReactionInfo_1 = require("./ReactionInfo");
/** Information about a Message's reaction. */
class MessageReactionInfo extends ReactionInfo_1.ReactionInfo {
    /**
     * @param data raw data.
     * @param client client.
     */
    constructor(data, client) {
        super(data, client);
        this.messageID = data.reaction.messageId;
        this.type = "message";
    }
    /** The message where the reaction has been added.
     * If the message is cached, it'll return a Message component,
     * otherwise it'll return basic information about this message.
     */
    get message() {
        const channel = this.client.getChannel(this.raw.serverId, this.raw.reaction.channelId);
        return channel?.messages?.get(this.messageID) ?? {
            id: this.messageID,
            guild: this.client.guilds.get(this.raw.serverId) ?? {
                id: this.raw.serverId
            },
            channelID: this.raw.reaction.channelId
        };
    }
}
exports.MessageReactionInfo = MessageReactionInfo;
