// touchguild/dist/lib/structures/Permission.js
"use strict";
/** @module Permission */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Permission = void 0;
/** Class representing a Permission */
class Permission {
    constructor(data) {
        this.data = data;
        this.createdAt = new Date(data.createdAt);
        this.updatedAt = data.updatedAt ?? null;
        this.target = data["userId"] ? "user" : (data["roleId"] ? "role" : "unknown");
        this.targetID = data["userId"] ?? data["roleId"] ?? null;
        this.parentID = data["channelId"] ?? data["categoryId"] ?? null;
        this.update(data);
    }
    has(...permissions) {
        for (const permission of permissions) {
            if (this.data.permissions[permission])
                return false;
        }
        return true;
    }
    toJSON() {
        return this.data.permissions;
    }
    toString() {
        let result = "";
        for (const permission of Object.entries(this.data.permissions)) {
            result += `${permission[0]} : ${permission[1].toString()}\n`;
        }
        return result;
    }
    update(data) {
        if (data.createdAt !== undefined) {
            this.createdAt = new Date(data.createdAt);
        }
        if (data.updatedAt !== undefined) {
            this.updatedAt = data.updatedAt ?? null;
        }
    }
}
exports.Permission = Permission;
