// touchguild/dist/lib/structures/ReactionInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReactionInfo = void 0;
/** Default information every other reaction has. */
class ReactionInfo {
    /**
     * @param data raw data.
     * @param client client.
     */
    constructor(data, client) {
        this.raw = data;
        this.channelID = data.reaction.channelId;
        this.reactorID = data.reaction.createdBy;
        this.emoji = {
            id: data.reaction.emote.id,
            name: data.reaction.emote.name,
            url: data.reaction.emote.url
        };
        Object.defineProperty(this, "client", {
            value: client,
            enumerable: false,
            writable: false,
            configurable: false
        });
    }
    /** Cached member. If member isn't cached will return an object with the member's id. */
    get reactor() {
        return this.client.getGuild(this.raw.serverId)?.members.get(this.raw.reaction.createdBy) ?? {
            id: this.raw.reaction.createdBy
        };
    }
}
exports.ReactionInfo = ReactionInfo;
