// touchguild/dist/lib/structures/SocialLink.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SocialLink = void 0;
/** User's social link. */
class SocialLink {
    /**
     * @param data raw data
     * @param client client
     */
    constructor(data, client) {
        this.client = client;
        this.type = data.type;
        this.userID = data.userId;
        this.handle = data.handle ?? null;
        this.serviceID = data.serviceId ?? null;
        this.createdAt = new Date(data.createdAt);
        this.update(data);
    }
    toJSON() {
        return {
            type: this.type,
            userID: this.userID,
            handle: this.handle,
            serviceID: this.serviceID,
            createdAt: this.createdAt
        };
    }
    update(data) {
        if (data.createdAt !== undefined) {
            this.createdAt = new Date(data.createdAt);
        }
        if (data.handle !== undefined) {
            this.handle = data.handle;
        }
        if (data.serviceId !== undefined) {
            this.serviceID = data.serviceId;
        }
        if (data.type !== undefined) {
            this.type = data.type;
        }
        if (data.userId !== undefined) {
            this.userID = data.userId;
        }
    }
    /** Retrieve cached user. */
    get user() {
        return this.client.users.get(this.userID) ?? null;
    }
}
exports.SocialLink = SocialLink;
