// touchguild/dist/lib/structures/TextChannel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TextChannel = void 0;
const tslib_1 = require("tslib");
const Message_1 = require("./Message");
const GuildChannel_1 = require("./GuildChannel");
const TypedCollection_1 = tslib_1.__importDefault(require("../util/TypedCollection"));
/** Represents a guild channel. */
class TextChannel extends GuildChannel_1.GuildChannel {
    /**
     * @param data raw data
     * @param client client
     */
    constructor(data, client) {
        super(data, client);
        this.messages = new TypedCollection_1.default(Message_1.Message, client, client.params.collectionLimits?.messages);
        this.update(data);
    }
    /** Create a message in this channel.
     * @param options Message options.
     */
    async createMessage(options) {
        return this.client.rest.channels.createMessage(this.id, options);
    }
    /** Edit a message from this channel.
     * @param options Message options.
     */
    async editMessage(messageID, options) {
        return this.client.rest.channels.editMessage(this.id, messageID, options);
    }
    /** Delete a message from this channel.
     * @param options Message options.
     */
    async deleteMessage(messageID) {
        return this.client.rest.channels.deleteMessage(this.id, messageID);
    }
    /** Create Channel Role Permissions
     * @param targetID ID of the target object (role or user) to assign the permission to.
     * @param options Create options.
     *
     * Warning: targetID must have the correct type (number=role, string=user).
     */
    async createPermission(targetID, options) {
        return this.client.rest.channels.createPermission(this.guildID, this.id, targetID, options);
    }
    /**
     * Edit a channel permission.
     * @param targetID ID of the target object (role or user) the permission is assigned to.
     * @param options Edit options
     *
     * Warning: targetID must have the correct type (number=role, string=user).
     */
    async editPermission(targetID, options) {
        return this.client.rest.channels.editPermission(this.guildID, this.id, targetID, options);
    }
    /**
     * Delete an existing permission set on this channel.
     * @param targetID ID of the target object (role or user) the permission is assigned to.
     *
     * Warning: targetID must have the correct type (number=role, string=user).
     */
    async deletePermission(targetID) {
        return this.client.rest.channels.deletePermission(this.guildID, this.id, targetID);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            messages: this.messages.map(message => message.toJSON())
        };
    }
}
exports.TextChannel = TextChannel;
