// touchguild/dist/lib/structures/UserClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserClient = void 0;
const User_1 = require("./User");
/** UserClient represents the logged bot's user. */
class UserClient extends User_1.User {
    /**
     * @param data raw data.
     * @param client client.
     */
    constructor(data, client) {
        super(data, client);
        this.botID = data.botId;
        this.type = "bot";
        this.bot = true;
        this.ownerID = data.createdBy;
        this.update(data);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            botID: this.botID,
            createdAt: this.createdAt,
            ownerID: this.ownerID
        };
    }
    update(data) {
        if (data.botId !== undefined) {
            this.botID = data.botId;
        }
        if (data.createdAt !== undefined) {
            this.createdAt = new Date(data.createdAt);
        }
        if (data.createdBy !== undefined) {
            this.ownerID = data.createdBy;
        }
        if (data.id !== undefined) {
            this.id = data.id;
        }
        if (data.name !== undefined) {
            this.username = data.name;
        }
        if (data.avatar !== undefined) {
            this.avatarURL = data.avatar;
        }
        if (data.banner !== undefined) {
            this.bannerURL = data.banner;
        }
        if (data.type !== undefined) {
            this.type = data.type;
        }
    }
}
exports.UserClient = UserClient;
