// touchguild/dist/lib/structures/User.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.User = void 0;
const Base_1 = require("./Base");
/** Represents a user. */
class User extends Base_1.Base {
    /**
     * @param data raw data.
     * @param client client.
     */
    constructor(data, client) {
        super(data.id, client);
        this.type = data.type ?? null;
        this.username = data.name;
        this.createdAt = new Date(data.createdAt);
        this.avatarURL = data.avatar ?? null;
        this.bannerURL = data.banner ?? null;
        if (!this.type)
            this.type = "user"; // since it's only defined when it's a bot.
        this.bot = this.type === "bot" ? true : false;
        this.update(data);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            type: this.type,
            username: this.username,
            createdAt: this.createdAt,
            avatarURL: this.avatarURL,
            bannerURL: this.bannerURL,
            bot: this.bot
        };
    }
    update(d) {
        const data = d;
        if (data.avatar !== undefined) {
            this.avatarURL = data.avatar ?? null;
        }
        if (data.banner !== undefined) {
            this.bannerURL = data.banner ?? null;
        }
        if (data.createdAt !== undefined) {
            this.createdAt = new Date(data.createdAt);
        }
        if (data.id !== undefined) {
            this.id = data.id;
        }
        if (data.name !== undefined) {
            this.username = data.name;
        }
        if (data.type !== undefined) {
            this.type = data.type ?? null;
        }
    }
}
exports.User = User;
