// touchguild/dist/lib/structures/Webhook.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Webhook = void 0;
const Base_1 = require("./Base");
/** Represents a Guild or channel webhook. */
class Webhook extends Base_1.Base {
    /**
     * @param data raw data.
     * @param client client.
     */
    constructor(data, client) {
        super(data.id, client);
        this.guildID = data.serverId;
        this.channelID = data.channelId;
        this.username = data.name;
        this.createdAt = new Date(data.createdAt);
        this.deletedAt = data.deletedAt ? new Date(data.deletedAt) : null;
        this.ownerID = data.createdBy;
        this.token = data.token ?? null;
        this.update(data);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            guildID: this.guildID,
            channelID: this.channelID,
            username: this.username,
            createdAt: this.createdAt,
            deletedAt: this.deletedAt,
            ownerID: this.ownerID,
            token: this.token
        };
    }
    update(data) {
        if (data.channelId !== undefined) {
            this.channelID = data.channelId;
        }
        if (data.createdAt !== undefined) {
            this.createdAt = new Date(data.createdAt);
        }
        if (data.createdBy !== undefined) {
            this.ownerID = data.createdBy;
        }
        if (data.deletedAt !== undefined) {
            this.deletedAt = new Date(data.deletedAt);
        }
        if (data.id !== undefined) {
            this.id = data.id;
        }
        if (data.name !== undefined) {
            this.username = data.name;
        }
        if (data.serverId !== undefined) {
            this.guildID = data.serverId;
        }
        if (data.token !== undefined) {
            this.token = data.token;
        }
    }
    /** Update the webhook. */
    async edit(options) {
        return this.client.rest.guilds.editWebhook(this.guildID, this.id, options);
    }
    /** Delete the webhook. */
    async delete() {
        return this.client.rest.guilds.deleteWebhook(this.guildID, this.id);
    }
}
exports.Webhook = Webhook;
