// touchguild/dist/lib/routes/Misc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Miscellaneous = void 0;
const tslib_1 = require("tslib");
const endpoints = tslib_1.__importStar(require("../rest/endpoints"));
const SocialLink_1 = require("../structures/SocialLink");
const Guild_1 = require("../structures/Guild");
/** Miscellaneous routes. */
class Miscellaneous {
    #manager;
    /**
     * @param manager REST Manager needed to execute request.
     */
    constructor(manager) {
        this.#manager = manager;
    }
    /** Get a specified social link from the member, if member is connected to them through Guilded.
     * @param guildID The ID of the guild the member is in.
     * @param memberID The ID of the member to get their social link.
     * @param socialMediaName Name of a social media linked to this member.
     */
    async getSocialLink(guildID, memberID, socialMediaName) {
        return this.#manager.authRequest({
            method: "GET",
            path: endpoints.GUILD_MEMBER_SOCIALS(guildID, memberID, socialMediaName)
        }).then(data => new SocialLink_1.SocialLink(data.socialLink, this.#manager.client));
    }
    /**
     * Get a user.
     *
     * Note: when getting the bot's user, only the information specific to 'User' will be returned.
     * If you'd like to get the UserClient (the bot itself), use Client#user.
     * @param userID The ID of the user to get.
     */
    async getUser(userID) {
        return this.#manager.authRequest({
            method: "GET",
            path: endpoints.USER(userID)
        }).then(data => this.#manager.client.util.updateUser(data.user));
    }
    /**
     * Retrieve user's joined servers.
     * @param userID ID of the user. (`@me` can be used to select your instance)
     */
    async getUserGuilds(userID) {
        return this.#manager.authRequest({
            method: "GET",
            path: endpoints.USER_SERVERS(userID)
        }).then(data => data.servers.map(d => new Guild_1.Guild(d, this.#manager.client)));
    }
    /**
     * Change a user's status, this includes the bot's one.
     * @param userID User ID (@me can be used).
     * @param options Status options
     */
    async updateUserStatus(userID, options) {
        return this.#manager.authRequest({
            method: "PUT",
            path: endpoints.USER_STATUS(userID),
            json: options
        });
    }
    /**
     * Delete a user's status, this includes the bot's one.
     * @param userID User ID (@me can be used).
     */
    async deleteUserStatus(userID) {
        return this.#manager.authRequest({
            method: "DELETE",
            path: endpoints.USER_STATUS(userID)
        });
    }
}
exports.Miscellaneous = Miscellaneous;
