// touchguild/dist/lib/gateway/events/ChannelHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChannelHandler = void 0;
/** @module ChannelHandler */
const GatewayEventHandler_1 = require("./GatewayEventHandler");
/** Internal component, emitting channel events. */
class ChannelHandler extends GatewayEventHandler_1.GatewayEventHandler {
    async channelCreate(data) {
        if (this.client.params.waitForCaching)
            await this.addGuildChannel(data.serverId, data.channel.id);
        else
            void this.addGuildChannel(data.serverId, data.channel.id);
        const ChannelComponent = this.client.util.updateChannel(data.channel);
        this.client.emit("channelCreate", ChannelComponent);
    }
    async channelUpdate(data) {
        if (this.client.params.waitForCaching)
            await this.addGuildChannel(data.serverId, data.channel.id);
        else
            void this.addGuildChannel(data.serverId, data.channel.id);
        const channel = this.client.getChannel(data.serverId, data.channel.id);
        const CachedChannel = channel ? channel.toJSON() : null;
        const ChannelComponent = this.client.util.updateChannel(data.channel);
        this.client.emit("channelUpdate", ChannelComponent, CachedChannel);
    }
    async channelDelete(data) {
        const guild = this.client.guilds.get(data.serverId);
        const ChannelComponent = this.client.util.updateChannel(data.channel);
        guild?.channels.delete(data.channel.id);
        this.client.emit("channelDelete", ChannelComponent);
    }
    async channelRolePermissionCreated(data) {
        const ChannelComponent = data.channelRolePermission;
        this.client.emit("channelRolePermissionCreated", ChannelComponent);
    }
    async channelRolePermissionUpdated(data) {
        const ChannelComponent = data.channelRolePermission;
        this.client.emit("channelRolePermissionUpdated", ChannelComponent);
    }
    async channelRolePermissionDeleted(data) {
        const ChannelComponent = data.channelRolePermission;
        this.client.emit("channelRolePermissionDeleted", ChannelComponent);
    }
    async channelUserPermissionCreated(data) {
        const ChannelComponent = data.channelUserPermission;
        this.client.emit("channelUserPermissionCreated", ChannelComponent);
    }
    async channelUserPermissionUpdated(data) {
        const ChannelComponent = data.channelUserPermission;
        this.client.emit("channelUserPermissionUpdated", ChannelComponent);
    }
    async channelUserPermissionDeleted(data) {
        const ChannelComponent = data.channelUserPermission;
        this.client.emit("channelUserPermissionDeleted", ChannelComponent);
    }
    async channelCategoryUserPermissionCreated(data) {
        const ChannelComponent = data.channelCategoryUserPermission;
        this.client.emit("channelCategoryUserPermissionCreated", ChannelComponent);
    }
    async channelCategoryUserPermissionUpdated(data) {
        const ChannelComponent = data.channelCategoryUserPermission;
        this.client.emit("channelCategoryUserPermissionUpdated", ChannelComponent);
    }
    async channelCategoryUserPermissionDeleted(data) {
        const ChannelComponent = data.channelCategoryUserPermission;
        this.client.emit("channelCategoryUserPermissionDeleted", ChannelComponent);
    }
    async channelCategoryRolePermissionCreated(data) {
        const ChannelComponent = data.channelCategoryRolePermission;
        this.client.emit("channelCategoryRolePermissionCreated", ChannelComponent);
    }
    async channelCategoryRolePermissionUpdated(data) {
        const ChannelComponent = data.channelCategoryRolePermission;
        this.client.emit("channelCategoryRolePermissionUpdated", ChannelComponent);
    }
    async channelCategoryRolePermissionDeleted(data) {
        const ChannelComponent = data.channelCategoryRolePermission;
        this.client.emit("channelCategoryRolePermissionDeleted", ChannelComponent);
    }
    async channelArchive(data) {
        const ChannelComponent = this.client.util.updateChannel(data.channel);
        this.client.emit("channelArchive", ChannelComponent);
    }
    async channelRestore(data) {
        const ChannelComponent = this.client.util.updateChannel(data.channel);
        this.client.emit("channelRestore", ChannelComponent);
    }
    async addGuildChannel(guildID, channelID) {
        if (this.client.getChannel(guildID, channelID) !== undefined)
            return;
        const channel = await this.client.rest.channels.getChannel(channelID).catch(err => this.client.emit("warn", `Cannot register channel to cache due to: (${String(err)})`));
        const guild = this.client.guilds.get(guildID);
        if (typeof channel !== "boolean")
            guild?.channels?.add(channel);
    }
}
exports.ChannelHandler = ChannelHandler;
