// touchguild/dist/lib/gateway/events/GuildHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GuildHandler = void 0;
/** @module GuildHandler */
const GatewayEventHandler_1 = require("./GatewayEventHandler");
const index_1 = require("../../index");
const MemberUpdateInfo_1 = require("../../structures/MemberUpdateInfo");
const MemberRemoveInfo_1 = require("../../structures/MemberRemoveInfo");
const GuildGroup_1 = require("../../structures/GuildGroup");
const GuildCategory_1 = require("../../structures/GuildCategory");
/** Internal component, emitting guild events. */
class GuildHandler extends GatewayEventHandler_1.GatewayEventHandler {
    guildBanAdd(data) {
        const GuildMemberBanComponent = new index_1.BannedMember(data.serverId, data.serverMemberBan, this.client);
        this.client.emit("guildBanAdd", GuildMemberBanComponent);
    }
    guildBanRemove(data) {
        const GuildMemberBanComponent = new index_1.BannedMember(data.serverId, data.serverMemberBan, this.client);
        this.client.emit("guildBanRemove", GuildMemberBanComponent);
    }
    guildMemberAdd(data) {
        const MemberComponent = new index_1.Member(data.member, this.client, data.serverId);
        this.client.emit("guildMemberAdd", MemberComponent, data.serverMemberCount);
    }
    guildMemberRemove(data) {
        const output = new MemberRemoveInfo_1.MemberRemoveInfo(data, data.userId, this.client);
        this.client.emit("guildMemberRemove", output);
    }
    guildMemberUpdate(data) {
        const output = new MemberUpdateInfo_1.MemberUpdateInfo(data, data.userInfo.id, this.client);
        this.client.emit("guildMemberUpdate", output);
    }
    guildMemberRoleUpdate(data) {
        const output = new MemberUpdateInfo_1.MemberUpdateInfo(data, data.memberRoleIds[0].userId, this.client);
        this.client.emit("guildMemberUpdate", output);
    }
    guildCreate(data) {
        const GuildComponent = new index_1.Guild(data.server, this.client);
        this.client.guilds.add(GuildComponent);
        const output = {
            guild: GuildComponent,
            inviterID: data.createdBy
        };
        this.client.emit("guildCreate", output);
    }
    guildDelete(data) {
        const GuildComponent = new index_1.Guild(data.server, this.client);
        this.client.guilds.delete(GuildComponent.id);
        const output = {
            guild: GuildComponent,
            removerID: data.createdBy
        };
        this.client.emit("guildDelete", output);
    }
    guildMemberSocialLinkCreate(data) {
        this.client.emit("guildMemberUpdate", new MemberUpdateInfo_1.MemberUpdateInfo(data, data.socialLink.userId, this.client));
    }
    guildMemberSocialLinkUpdate(data) {
        this.client.emit("guildMemberUpdate", new MemberUpdateInfo_1.MemberUpdateInfo(data, data.socialLink.userId, this.client));
    }
    guildMemberSocialLinkDelete(data) {
        this.client.emit("guildMemberUpdate", new MemberUpdateInfo_1.MemberUpdateInfo(data, data.socialLink.userId, this.client));
    }
    guildGroupCreate(data) {
        const GuildGroupComponent = new GuildGroup_1.GuildGroup(data.group, this.client);
        this.client.guilds.get(data.serverId)?.groups.add(GuildGroupComponent);
        this.client.emit("guildGroupCreate", GuildGroupComponent);
    }
    guildGroupUpdate(data) {
        const guild = this.client.guilds.get(data.serverId);
        const CachedGroup = guild?.groups.get(data.group.id)?.toJSON() ?? null;
        const GuildGroupComponent = guild?.groups.update(new GuildGroup_1.GuildGroup(data.group, this.client)) ?? new GuildGroup_1.GuildGroup(data.group, this.client);
        this.client.emit("guildGroupUpdate", GuildGroupComponent, CachedGroup);
    }
    guildGroupDelete(data) {
        const guild = this.client.guilds.get(data.serverId);
        const GuildGroupComponent = guild?.groups.update(new GuildGroup_1.GuildGroup(data.group, this.client)) ?? new GuildGroup_1.GuildGroup(data.group, this.client);
        this.client.emit("guildGroupDelete", GuildGroupComponent);
    }
    guildRoleCreate(data) {
        const guild = this.client.guilds.get(data.serverId);
        const role = guild?.roles.add(new index_1.GuildRole(data.role, this.client)) ?? new index_1.GuildRole(data.role, this.client);
        this.client.emit("guildRoleCreate", role);
    }
    guildRoleUpdate(data) {
        const guild = this.client.guilds.get(data.serverId);
        const cachedRole = guild?.roles.get(data.role.id)?.toJSON() ?? null;
        const role = guild?.roles.update(new index_1.GuildRole(data.role, this.client)) ?? new index_1.GuildRole(data.role, this.client);
        this.client.emit("guildRoleUpdate", role, cachedRole);
    }
    guildRoleDelete(data) {
        const guild = this.client.guilds.get(data.serverId);
        const role = guild?.roles.update(new index_1.GuildRole(data.role, this.client)) ?? new index_1.GuildRole(data.role, this.client);
        guild?.roles.delete(data.role.id);
        this.client.emit("guildRoleDelete", role);
    }
    guildCategoryCreate(data) {
        const category = new GuildCategory_1.GuildCategory(data.category, this.client);
        this.client.emit("guildCategoryCreate", category);
    }
    guildCategoryUpdate(data) {
        const category = new GuildCategory_1.GuildCategory(data.category, this.client);
        this.client.emit("guildCategoryUpdate", category);
    }
    guildCategoryDelete(data) {
        const category = new GuildCategory_1.GuildCategory(data.category, this.client);
        this.client.emit("guildCategoryDelete", category);
    }
}
exports.GuildHandler = GuildHandler;
