// touchguild/dist/lib/gateway/events/ListItemHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListItemHandler = void 0;
/** @module ListItemHandler */
const GatewayEventHandler_1 = require("./GatewayEventHandler");
const ListItem_1 = require("../../structures/ListItem");
/** Internal component, emitting list events. */
class ListItemHandler extends GatewayEventHandler_1.GatewayEventHandler {
    listItemCreate(data) {
        const ListItemComponent = new ListItem_1.ListItem(data.listItem, this.client);
        this.client.emit("listItemCreate", ListItemComponent);
    }
    listItemUpdate(data) {
        const ListItemComponent = new ListItem_1.ListItem(data.listItem, this.client);
        this.client.emit("listItemUpdate", ListItemComponent);
    }
    listItemDelete(data) {
        const ListItemComponent = new ListItem_1.ListItem(data.listItem, this.client);
        this.client.emit("listItemDelete", ListItemComponent);
    }
    listItemComplete(data) {
        const ListItemComponent = new ListItem_1.ListItem(data.listItem, this.client);
        this.client.emit("listItemComplete", ListItemComponent);
    }
    listItemUncomplete(data) {
        const ListItemComponent = new ListItem_1.ListItem(data.listItem, this.client);
        this.client.emit("listItemUncomplete", ListItemComponent);
    }
}
exports.ListItemHandler = ListItemHandler;
