// touchguild/dist/lib/gateway/events/MessageHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageHandler = void 0;
/** @module MessageHandler */
const GatewayEventHandler_1 = require("./GatewayEventHandler");
const Message_1 = require("../../structures/Message");
const MessageReactionInfo_1 = require("../../structures/MessageReactionInfo");
/** Internal component, emitting message events. */
class MessageHandler extends GatewayEventHandler_1.GatewayEventHandler {
    async messageCreate(data) {
        if (this.client.params.waitForCaching)
            await this.addGuildChannel(data.serverId, data.message.channelId);
        else
            void this.addGuildChannel(data.serverId, data.message.channelId);
        const channel = this.client.getChannel(data.serverId, data.message.channelId);
        const MessageComponent = channel?.messages?.update(data.message) ?? new Message_1.Message(data.message, this.client);
        this.client.emit("messageCreate", MessageComponent);
    }
    async messageUpdate(data) {
        if (this.client.params.waitForCaching)
            await this.addGuildChannel(data.serverId, data.message.channelId);
        else
            void this.addGuildChannel(data.serverId, data.message.channelId);
        const channel = this.client.getChannel(data.serverId, data.message.channelId);
        const CachedMessage = channel?.messages?.get(data.message.id)?.toJSON() ?? null;
        const MessageComponent = channel?.messages?.update(data.message) ?? new Message_1.Message(data.message, this.client);
        this.client.emit("messageUpdate", MessageComponent, CachedMessage);
    }
    async messageDelete(data) {
        if (this.client.params.waitForCaching)
            await this.addGuildChannel(data.serverId, data.message.channelId);
        else
            void this.addGuildChannel(data.serverId, data.message.channelId);
        const channel = this.client.getChannel(data.serverId, data.message.channelId);
        const PU_Message = channel?.messages.update(data.message) ?? {
            id: data.message.id,
            guildID: data.serverId,
            channelID: data.message.channelId,
            deletedAt: new Date(data.message.deletedAt),
            isPrivate: data.message.isPrivate ?? null
        };
        channel?.messages?.delete(data.message.id);
        this.client.emit("messageDelete", PU_Message);
    }
    async messageReactionAdd(data) {
        if (data.serverId)
            if (this.client.params.waitForCaching)
                await this.addGuildChannel(data.serverId, data.reaction.channelId);
            else
                void this.addGuildChannel(data.serverId, data.reaction.channelId);
        const ReactionInfo = new MessageReactionInfo_1.MessageReactionInfo(data, this.client);
        this.client.emit("reactionAdd", ReactionInfo);
    }
    async messageReactionRemove(data) {
        if (data.serverId)
            if (this.client.params.waitForCaching)
                await this.addGuildChannel(data.serverId, data.reaction.channelId);
            else
                void this.addGuildChannel(data.serverId, data.reaction.channelId);
        const ReactionInfo = new MessageReactionInfo_1.MessageReactionInfo(data, this.client);
        this.client.emit("reactionRemove", ReactionInfo);
    }
    async messageReactionBulkRemove(data) {
        if (data.serverId)
            if (this.client.params.waitForCaching)
                await this.addGuildChannel(data.serverId, data.channelId);
            else
                void this.addGuildChannel(data.serverId, data.channelId);
        const BulkRemoveInfo = {
            channelID: data.channelId,
            guildID: data.serverId,
            messageID: data.messageId,
            count: data.count,
            deletedBy: data.deletedBy,
            emote: data.emote ?? null
        };
        this.client.emit("reactionBulkRemove", BulkRemoveInfo);
    }
    async messagePin(data) {
        if (this.client.params.waitForCaching)
            await this.addGuildChannel(data.serverId, data.message.channelId);
        else
            void this.addGuildChannel(data.serverId, data.message.channelId);
        const channel = this.client.getChannel(data.serverId, data.message.channelId);
        const MessageComponent = channel?.messages?.update(data.message) ?? new Message_1.Message(data.message, this.client);
        this.client.emit("messagePin", MessageComponent);
    }
    async messageUnpin(data) {
        if (this.client.params.waitForCaching)
            await this.addGuildChannel(data.serverId, data.message.channelId);
        else
            void this.addGuildChannel(data.serverId, data.message.channelId);
        const channel = this.client.getChannel(data.serverId, data.message.channelId);
        const MessageComponent = channel?.messages?.update(data.message) ?? new Message_1.Message(data.message, this.client);
        this.client.emit("messageUnpin", MessageComponent);
    }
    async addGuildChannel(guildID, channelID) {
        if (this.client.getChannel(guildID, channelID) !== undefined)
            return;
        const channel = await this.client.rest.channels.getChannel(channelID).catch(err => this.client.emit("warn", `Cannot register channel to cache due to: (${String(err)})`));
        const guild = this.client.guilds.get(guildID);
        if (typeof channel !== "boolean")
            guild?.channels?.add(channel);
    }
}
exports.MessageHandler = MessageHandler;
