// touchguild/dist/lib/gateway/events/UserHandler.js
"use strict";
/** @module UserHandler */
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserHandler = void 0;
const GatewayEventHandler_1 = require("./GatewayEventHandler");
/** Internal component, emitting user events. */
class UserHandler extends GatewayEventHandler_1.GatewayEventHandler {
    userStatusCreate(data) {
        const userStatus = {
            content: data.userStatus.content ?? null,
            emoteID: data.userStatus.emoteId
        };
        const result = {
            expiresAt: data.expiresAt ?? null,
            userID: data.userId,
            userStatus
        };
        this.client.emit("userStatusCreate", result);
    }
    userStatusDelete(data) {
        const userStatus = {
            content: data.userStatus.content ?? null,
            emoteID: data.userStatus.emoteId
        };
        const result = {
            userID: data.userId,
            userStatus
        };
        this.client.emit("userStatusDelete", result);
    }
}
exports.UserHandler = UserHandler;
