// touchguild/dist/lib/gateway/WSManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WSManager = void 0;
const tslib_1 = require("tslib");
/** @module WSManager */
const GatewayError_1 = tslib_1.__importDefault(require("./GatewayError"));
const Constants_1 = require("../Constants");
const TypedEmitter_1 = tslib_1.__importDefault(require("../types/TypedEmitter"));
const pkgconfig_1 = require("../../pkgconfig");
const Util_1 = require("../util/Util");
const ws_1 = tslib_1.__importDefault(require("ws"));
const node_assert_1 = tslib_1.__importDefault(require("node:assert"));
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
let Erlpack;
try {
    // eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-unsafe-assignment, unicorn/prefer-module
    Erlpack = require("erlpack");
}
catch { }
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
let ZlibSync, zlibConstants;
try {
    // eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-unsafe-assignment, unicorn/prefer-module
    ZlibSync = require("zlib-sync");
    // eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-unsafe-assignment, unicorn/prefer-module
    zlibConstants = require("zlib-sync");
}
catch {
    try {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, unicorn/prefer-module
        ZlibSync = require("pako");
        // eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access, unicorn/prefer-module
        zlibConstants = require("pako").constants;
    }
    catch { }
}
/** Websocket manager, used to receive ws events. */
class WSManager extends TypedEmitter_1.default {
    #heartbeatInterval;
    #connectTimeout;
    #sharedZLib;
    constructor(client, params) {
        super();
        Object.defineProperties(this, {
            client: {
                value: client,
                enumerable: false,
                writable: false,
                configurable: false
            },
            ws: {
                value: null,
                enumerable: false,
                writable: true,
                configurable: false
            }
        });
        this.params = params;
        this.token = params.token;
        this.apiversion = params.apiversion ?? pkgconfig_1.config.GuildedAPI.GatewayVersion ?? 1;
        this.proxyURL = params.proxyURL ?? pkgconfig_1.config.GuildedAPI.GatewayURL ?? `wss://www.guilded.gg/websocket/v${this.apiversion}`;
        this.reconnect = params.reconnect ?? true;
        this.reconnectAttemptLimit = params.reconnectAttemptLimit ?? 1;
        this.reconnectInterval = 1000;
        this.replayMissedEvents = params.replayMissedEvents ?? true;
        this.#heartbeatInterval = null;
        this.ws = null;
        this.firstwsMessage = true;
        this.lastMessageID = undefined;
        this.currReconnectAttempt = 0;
        this.alive = false;
        this.latency = 1000;
        this.lastHeartbeatSent = NaN;
        this.lastHeartbeatReceived = NaN;
        this.lastHeartbeatAck = false;
        this.heartbeatRequested = false;
        this.connected = false;
        this.connectionTimeout = 30000;
        this.#connectTimeout = null;
        this.compression = params.compression ?? false; // not enabled, guilded doesn't support compression for now.
    }
    // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
    get vAPI() {
        return this.apiversion;
    }
    _debug(message) {
        return this.emit("debug", `[TouchGuild DEBUG]: ${message.toString()}`);
    }
    get replayEventsCondition() {
        return this.replayMissedEvents === true && this.lastMessageID !== undefined;
    }
    connect() {
        if (this.ws && this.ws.readyState !== ws_1.default.CLOSED) {
            this.client.emit("error", new Error("Calling connect while an existing connection is already established."));
            return;
        }
        this.currReconnectAttempt++;
        this.initialize();
    }
    initialize() {
        if (!this.token)
            return this.disconnect(false, new Error("Invalid Token."));
        if (this.compression) {
            if (!ZlibSync) {
                throw new Error("Cannot use compression without the pako or zlib-sync module.");
            }
            this.client.emit("debug", "Initializing compression.");
            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call
            this.#sharedZLib = new ZlibSync.Inflate({ chunkSize: 128 * 1024 });
        }
        const wsoptions = { headers: { Authorization: `Bearer ${this.params.token}` }, protocol: "HTTPS" };
        if (this.replayEventsCondition)
            Object.assign(wsoptions.headers, { "guilded-last-message-id": this.lastMessageID });
        this.ws = new ws_1.default(this.proxyURL, wsoptions);
        this.ws.on("open", this.onSocketOpen.bind(this));
        this.ws.on("close", this.onSocketClose.bind(this));
        this.ws.on("ping", this.onSocketPing.bind(this));
        this.ws.on("pong", this.onSocketPong.bind(this));
        this.ws.on("message", (args) => {
            if (this.firstwsMessage === true) {
                this.firstwsMessage = false;
            }
            this.onSocketMessage(args);
        });
        this.ws.on("error", (err) => {
            this.onSocketError.bind(this)(err);
            console.error("GATEWAY ERR: Couldn't connect to Guilded.");
        });
        this.#connectTimeout = setTimeout(() => {
            if (!this.connected) {
                this.disconnect(undefined, new Error("Connection timeout."));
            }
        }, this.connectionTimeout);
    }
    onSocketMessage(data) {
        if (typeof data === "string") {
            data = Buffer.from(data);
        }
        try {
            if (data instanceof ArrayBuffer) {
                if (this.compression || Erlpack) {
                    data = Buffer.from(data);
                }
            }
            else if (Array.isArray(data)) {
                data = Buffer.concat(data);
            }
            (0, node_assert_1.default)((0, Util_1.is)(data));
            if (this.compression) {
                if (data.length >= 4 && data.readUInt32BE(data.length - 4) === 0x30307D7D) {
                    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
                    this.#sharedZLib.push(data, zlibConstants.Z_SYNC_FLUSH);
                    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
                    if (this.#sharedZLib.err) {
                        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/restrict-template-expressions
                        this.client.emit("error", new Error(`ZLib ERROR ${this.#sharedZLib.err}: ${this.#sharedZLib.msg ?? ""}`));
                        return;
                    }
                    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-argument
                    data = Buffer.from(this.#sharedZLib.result ?? "");
                    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
                    return this.onPacket((Erlpack ? Erlpack.unpack(data) : JSON.parse(data.toString())));
                }
                else {
                    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
                    this.#sharedZLib.push(data, false);
                }
            }
            else if (Erlpack) {
                // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
                return this.onPacket(Erlpack.unpack(data));
            }
            else {
                return this.onPacket(JSON.parse(data.toString()));
            }
        }
        catch (err) {
            this.client.emit("error", err);
        }
    }
    onPacket(packet) {
        // s: Message ID used for replaying events after a disconnect.
        if (packet.s)
            this.lastMessageID = packet.s;
        switch (packet.op) {
            case Constants_1.GatewayOPCodes.Event: {
                this.emit("GATEWAY_PARSED_PACKET", packet.t, packet.d);
                this.emit("GATEWAY_PACKET", packet);
                break;
            }
            case Constants_1.GatewayOPCodes.Welcome: {
                if (!packet.d)
                    throw new Error("WSERR: Couldn't get packet data.");
                if (!packet.d["heartbeatIntervalMs"])
                    throw new Error("WSERR: Couldn't get the heartbeat interval.");
                if (this.#connectTimeout) {
                    clearInterval(this.#connectTimeout);
                }
                this.#heartbeatInterval = setInterval(() => this.heartbeat(), packet.d["heartbeatIntervalMs"]);
                this.emit("GATEWAY_WELCOME", packet.d["user"]);
                this.emit("GATEWAY_WELCOME_PACKET", packet);
                this.connected = true;
                break;
            }
            case Constants_1.GatewayOPCodes.Resume: {
                this.lastMessageID = undefined;
                break;
            }
            default: {
                this.emit("GATEWAY_UNKNOWN_PACKET", "??UNKNOWN OPCODE??", packet);
                break;
            }
        }
    }
    onSocketOpen() {
        this.alive = true;
        this.emit("debug", "Socket connection is open.");
    }
    onSocketError(error) {
        this.client.emit("error", error);
        this.emit("error", error);
        this.emit("exit", error);
        this.alive = false;
        return void 0;
    }
    onSocketClose(code, r) {
        const reason = r.toString();
        let reconnect;
        let err;
        this.alive = false;
        if (code) {
            this.client.emit("debug", `${code === 1000 ? "Clean" : "Unclean"} WS close: ${code}: ${reason}`);
            switch (code) {
                case 1006: {
                    err = new GatewayError_1.default("Connection lost", code);
                    break;
                }
                default: {
                    err = new GatewayError_1.default(reason, code);
                    break;
                }
            }
            this.disconnect(reconnect, err);
        }
    }
    onSocketPing() {
        // this._debug("Heartbeat has been sent.");
        this.ws.ping();
        this.lastHeartbeatSent = Date.now();
    }
    onSocketPong() {
        this.client.emit("debug", "Heartbeat acknowledged.");
        this.latency = this.lastHeartbeatSent - Date.now();
        this.lastHeartbeatAck = true;
    }
    heartbeat() {
        if (this.heartbeatRequested) {
            if (!this.lastHeartbeatAck) {
                this.lastHeartbeatAck = false;
                return this.client.emit("error", new Error("Server didn't acknowledge the previous heartbeat, possible lost connection."));
            }
            this.heartbeatRequested = false;
        }
        else {
            this.client.emit("debug", "Heartbeat requested.");
            this.ws?.ping();
            this.heartbeatRequested = true;
            this.lastHeartbeatAck = false;
        }
    }
    disconnect(reconnect = this.reconnect, error) {
        this.ws?.close();
        this.alive = false;
        this.connected = false;
        if (this.#heartbeatInterval) {
            clearInterval(this.#heartbeatInterval);
            this.#heartbeatInterval = null;
        }
        if (this.ws?.readyState !== ws_1.default.CLOSED) {
            this.ws?.removeAllListeners();
            try {
                if (reconnect) {
                    if (this.ws?.readyState !== ws_1.default.OPEN) {
                        this.ws?.close(4999, "Reconnect");
                    }
                    else {
                        this.client.emit("debug", "Closing websocket.");
                        this.ws.terminate();
                    }
                }
                else {
                    this.ws?.close(1000, "Normal Close");
                }
            }
            catch (err) {
                this.client.emit("error", err);
            }
        }
        if (error) {
            if (error instanceof GatewayError_1.default && [1001, 1006].includes(error.code)) {
                this.client.emit("debug", error.message);
            }
            else {
                this.client.emit("error", error);
            }
        }
        this.ws = null;
        this.reset();
        this.emit("disconnect", error);
        if (this.currReconnectAttempt >= this.reconnectAttemptLimit) {
            this.client.emit("debug", `Automatically invalidating session due to excessive resume attempts | Attempt ${this.currReconnectAttempt}`);
        }
        if (reconnect) {
            if (this.lastMessageID) {
                this.client.emit("debug", `Immediately reconnecting for potential resume | Attempt ${this.currReconnectAttempt}`);
                this.connect();
            }
            else {
                this.client.emit("debug", `Queueing reconnect in ${this.reconnectInterval}ms | Attempt ${this.currReconnectAttempt}`);
                setTimeout(() => {
                    this.connect();
                }, this.reconnectInterval);
                this.reconnectInterval = Math.min(Math.round(this.reconnectInterval * (Math.random() * 2 + 1)), 30000);
            }
        }
        else {
            this.hardReset();
        }
    }
    reset() {
        this.ws = null;
        this.firstwsMessage = true;
        this.lastMessageID = undefined;
        this.currReconnectAttempt = 0;
        this.alive = false;
        this.latency = Infinity;
        this.lastHeartbeatSent = NaN;
        this.lastHeartbeatReceived = NaN;
        this.lastHeartbeatAck = false;
        this.heartbeatRequested = false;
        this.connectionTimeout = 30000;
        this.#connectTimeout = null;
    }
    hardReset() {
        this.reset();
        this.token = this.params.token;
        this.apiversion = this.params.apiversion ?? pkgconfig_1.config.GuildedAPI.GatewayVersion ?? 1;
        this.proxyURL = this.params.proxyURL ?? pkgconfig_1.config.GuildedAPI.GatewayURL ?? `wss://www.guilded.gg/websocket/v${this.apiversion}`;
        this.reconnect = this.params.reconnect ?? true;
        this.reconnectAttemptLimit = this.params.reconnectAttemptLimit ?? 1;
        this.replayMissedEvents = this.params.replayMissedEvents ?? true;
        this.#heartbeatInterval = null;
        this.ws = null;
        this.firstwsMessage = true;
        this.lastMessageID = undefined;
        this.currReconnectAttempt = 0;
        this.alive = false;
        this.latency = Infinity;
        this.lastHeartbeatSent = NaN;
        this.lastHeartbeatReceived = NaN;
        this.lastHeartbeatAck = false;
        this.heartbeatRequested = false;
        this.connectionTimeout = 30000;
        this.#connectTimeout = null;
    }
}
exports.WSManager = WSManager;
